<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en" dir="ltr">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
	<!-- headlinks removed -->
	<link rel="shortcut icon" href="../../../../misc/favicon.ico"/>
    <title>Keyword Example Scripts - Phaserwiki</title>
    <style type="text/css">/*<![CDATA[*/ @import "../../../../skins/offline/main.css"; /*]]>*/</style>
    <link rel="stylesheet" type="text/css" media="print" href="../../../../skins/common/commonPrint.css" />
    <!--[if lt IE 5.5000]><style type="text/css">@import "../../../../skins/monobook/IE50Fixes.css";</style><![endif]-->
    <!--[if IE 5.5000]><style type="text/css">@import "../../../../skins/monobook/IE55Fixes.css";</style><![endif]-->
    <!--[if IE 6]><style type="text/css">@import "../../../../skins/monobook/IE60Fixes.css";</style><![endif]-->
    <!--[if IE]><script type="text/javascript" src="../../../../skins/common/IEFixes.js"></script>
    <meta http-equiv="imagetoolbar" content="no" /><![endif]-->
    <script type="text/javascript" src="../../../../skins/common/wikibits.js"></script>
    <script type="text/javascript" src="../../../../skins/offline/md5.js"></script>
    <script type="text/javascript" src="../../../../skins/offline/utf8.js"></script>
    <script type="text/javascript" src="../../../../skins/offline/lookup.js"></script>
    <script type="text/javascript" src="../../../../raw/gen.js"></script>                  </head>
  <body
    class="ns-0 ns-subject page-Keyword_Example_Scripts">
    <div id="globalWrapper">
      <div id="column-content">
	<div id="content">
	  <a name="top" id="contentTop"></a>
	        <h1 class="firstHeading">Keyword Example Scripts</h1>
	  <div id="bodyContent">
	    <h3 id="siteSub">From Phaserwiki</h3>
	    <div id="contentSub"></div>
	    	    	    <!-- start content -->
	    <div style="margin-left: 25px; float: right;"><table id="toc" class="toc" summary="Contents"><tr><td><div id="toctitle"><h2>Contents</h2></div>
<ul>
<li class="toclevel-1"><a href="#How_to_Define_Models"><span class="tocnumber">1</span> <span class="toctext">How to Define Models</span></a>
<ul>
<li class="toclevel-2"><a href="#Building_an_Ensemble_from_Coordinates"><span class="tocnumber">1.1</span> <span class="toctext">Building an Ensemble from Coordinates</span></a></li>
<li class="toclevel-2"><a href="#Building_an_Ensemble_from_Electron_Density"><span class="tocnumber">1.2</span> <span class="toctext">Building an Ensemble from Electron Density</span></a></li>
</ul>
</li>
<li class="toclevel-1"><a href="#How_to_Define_Composition"><span class="tocnumber">2</span> <span class="toctext">How to Define Composition</span></a>
<ul>
<li class="toclevel-2"><a href="#Composition_by_Molecular_Weight"><span class="tocnumber">2.1</span> <span class="toctext">Composition by Molecular Weight</span></a></li>
<li class="toclevel-2"><a href="#Composition_by_Sequence"><span class="tocnumber">2.2</span> <span class="toctext">Composition by Sequence</span></a></li>
<li class="toclevel-2"><a href="#Composition_by_Percentage_Scattering"><span class="tocnumber">2.3</span> <span class="toctext">Composition by Percentage Scattering</span></a></li>
</ul>
</li>
<li class="toclevel-1"><a href="#How_to_Define_Solutions"><span class="tocnumber">3</span> <span class="toctext">How to Define Solutions</span></a>
<ul>
<li class="toclevel-2"><a href="#.22sol.22_Files"><span class="tocnumber">3.1</span> <span class="toctext">"sol" Files</span></a></li>
<li class="toclevel-2"><a href="#.22rlist.22_Files"><span class="tocnumber">3.2</span> <span class="toctext">"rlist" Files</span></a></li>
<li class="toclevel-2"><a href="#Fixed_Partial_Structure"><span class="tocnumber">3.3</span> <span class="toctext">Fixed Partial Structure</span></a></li>
</ul>
</li>
<li class="toclevel-1"><a href="#Automated_Molecular_Replacement"><span class="tocnumber">4</span> <span class="toctext">Automated Molecular Replacement</span></a></li>
<li class="toclevel-1"><a href="#Non-Automated_Molecular_Replacement"><span class="tocnumber">5</span> <span class="toctext">Non-Automated Molecular Replacement</span></a>
<ul>
<li class="toclevel-2"><a href="#Fast_Rotation_Function"><span class="tocnumber">5.1</span> <span class="toctext">Fast Rotation Function</span></a></li>
<li class="toclevel-2"><a href="#Brute_Rotation_Function"><span class="tocnumber">5.2</span> <span class="toctext">Brute Rotation Function</span></a></li>
<li class="toclevel-2"><a href="#Fast_Translation_Function"><span class="tocnumber">5.3</span> <span class="toctext">Fast Translation Function</span></a></li>
<li class="toclevel-2"><a href="#Brute_Translation_Function"><span class="tocnumber">5.4</span> <span class="toctext">Brute Translation Function</span></a></li>
<li class="toclevel-2"><a href="#Refinement_and_Phasing"><span class="tocnumber">5.5</span> <span class="toctext">Refinement and Phasing</span></a></li>
<li class="toclevel-2"><a href="#Log-Likelihood_Gain"><span class="tocnumber">5.6</span> <span class="toctext">Log-Likelihood Gain</span></a></li>
<li class="toclevel-2"><a href="#Packing"><span class="tocnumber">5.7</span> <span class="toctext">Packing</span></a></li>
</ul>
</li>
<li class="toclevel-1"><a href="#Automated_Experimental_Phasing"><span class="tocnumber">6</span> <span class="toctext">Automated Experimental Phasing</span></a></li>
<li class="toclevel-1"><a href="#Anisotropy_Correction"><span class="tocnumber">7</span> <span class="toctext">Anisotropy Correction</span></a></li>
<li class="toclevel-1"><a href="#Cell_Content_Analysis"><span class="tocnumber">8</span> <span class="toctext">Cell Content Analysis</span></a></li>
<li class="toclevel-1"><a href="#Translational_NCS_Analysis"><span class="tocnumber">9</span> <span class="toctext">Translational NCS Analysis</span></a></li>
<li class="toclevel-1"><a href="#Normal_Mode_Analysis"><span class="tocnumber">10</span> <span class="toctext">Normal Mode Analysis</span></a></li>
</ul>
</td></tr></table><script type="text/javascript"> if (window.showTocToggle) { var tocShowText = "show"; var tocHideText = "hide"; showTocToggle(); } </script>
</div>
<p>Phaser can be run from the command line 
</p>
<dl><dt>$prompt&gt; phaser
</dt></dl>
<p>After printing the Phaser Banner, the Preprocessor will ask:
</p>
<dl><dt>ENTER KEYWORD INPUT FROM FILE OR FROM STANDARD INPUT
</dt></dl>
<p>Enter the keyword input
</p><p>At the end of the input, start Phaser with one of the commands
</p>
<ul><li>END
</li><li>QUIT
</li><li>STOP
</li><li>KILL
</li><li>EXIT
</li><li>GO
</li><li>RUN
</li><li>START
</li></ul>
<p>Alternatively, phaser can be run from command scripts as below
</p><p><br />
</p>
<a name="How_to_Define_Models" id="How_to_Define_Models"></a><h2> <span class="mw-headline">How to Define Models</span></h2>
<a name="Building_an_Ensemble_from_Coordinates" id="Building_an_Ensemble_from_Coordinates"></a><h3> <span class="mw-headline">Building an Ensemble from Coordinates</span></h3>
<p>You have one structure as a model with 44% sequence identity to the protein in the crystal.
</p>
<pre>
  ENSEmble mol1 PDB structure1.pdb IDENtity .44
</pre>
<p>You have three structures as models with 44%, 39% and 35% identity to the protein in the crystal.
</p>
<pre>
  ENSEmble mol2   PDB structure1.pdb IDENtity .44 PDB structure2.pdb IDENtity .39 PDB structure3.pdb IDENtity .35
</pre>
<p>You have an NMR Ensemble as a model. There is no need to split the coordinates in the pdb file provided that the models are separated by MODEL and ENDMDL cards. In this case the sequence identity is not a good indication of the rms deviation of the structural coordinates to the target structure. You should use the RMS option; several test cases have succeeded where the ID was close to 100% with an RMS value of about 1.5Å (see table below).
</p>
<pre>
  ENSEmble mol3 PDB nmr.pdb RMS 1.5
</pre>
<a name="Building_an_Ensemble_from_Electron_Density" id="Building_an_Ensemble_from_Electron_Density"></a><h3> <span class="mw-headline">Building an Ensemble from Electron Density</span></h3>
<p>You have low resolution electron density of your model. This density has been cut out and converted to structure factors in a large cell.
</p>
<pre>
  ENSEmble mol1  HKLIn mol1.mtz F = Fmol1 P = Pmol1 EXTEnt 23 25 29 RMS 2.0 CENTre 4 3 30 PROTein MW 10241 NUCLeic MW 0
</pre>
<a name="How_to_Define_Composition" id="How_to_Define_Composition"></a><h2> <span class="mw-headline">How to Define Composition</span></h2>
<a name="Composition_by_Molecular_Weight" id="Composition_by_Molecular_Weight"></a><h3> <span class="mw-headline">Composition by Molecular Weight</span></h3>
<p>You have one protein (with MW 21022) in the asymmetric unit
</p>
<pre>
COMPosition PROTein MW 21022
</pre>
<p>You have three copies of a protein (with MW 21022) in the asymmetric unit
</p>
<pre>
  COMPosition PROTein MW 21022
  COMPosition PROTein MW 21022
  COMPosition PROTein MW 21022
</pre>
<p>Another way of entering the same thing is
</p>
<pre>
  COMPosition PROTein MW 21022 NUMber 3
</pre>
<p>Yet another way of entering the same thing is
</p>
<pre>
  COMPosition PROTein MW 63066
</pre>
<p>You have two copies of a protein (with MW 21022), two copies of a protein (with MW 9843) and RNA with (MW 32004) in the asymmetric unit
</p>
<pre>
  COMPosition PROTein MW 21022 NUMber 2
  COMPosition PROTein MW 9843 NUMber 2
  COMPosition NUCLeic MW 32004
</pre>
<a name="Composition_by_Sequence" id="Composition_by_Sequence"></a><h3> <span class="mw-headline">Composition by Sequence</span></h3>
<p>You have one protein (with sequence in fasta format in the file prot1.seq) in the asymmetric unit
</p>
<pre>
  COMPosition PROTein SEQuence prot1.seq
</pre>
<p>You have three copies of a protein (with sequence in fasta format in the file prot1.seq) in the asymmetric unit
</p>
<pre>
  COMPosition PROTein SEQuence prot1.seq
  COMPosition PROTein SEQuence prot1.seq
  COMPosition PROTein SEQuence prot1.seq
</pre>
<p>Another way of entering the same thing is
</p>
<pre>
  COMPosition PROTein SEQuence prot1.seq NUMber 3
</pre>
<p>Yet another way of entering the same thing is to make a sequence file with all the amino acids concatenated together (prot1.seq3)
</p>
<pre>
  COMPosition PROTein SEQuence prot1.seq3
</pre>
<p>You have two copies of a protein (with sequence in fasta format in the file prot1.seq), two copies of a protein (with sequence in fasta format in the file prot2.seq) and RNA with (with sequence in fasta format in the file nucl1.seq) in the asymmetric unit
</p>
<pre>
  COMPosition PROTein SEQuence prot1.seq NUMber 2
  COMPosition PROTein SEQuence prot2.seq NUMber 2
  COMPosition NUCLeic SEQuence nucl1.seq
</pre>
<a name="Composition_by_Percentage_Scattering" id="Composition_by_Percentage_Scattering"></a><h3> <span class="mw-headline">Composition by Percentage Scattering</span></h3>
<p>Each copy of Ensemble mol1 gives 22% of the scattering
</p>
<pre>
  COMPosition ENSEmble mol1 FRACtional 0.22
</pre>
<p>Each copy of Ensemble mol2 gives 78% of the scattering
</p>
<pre>
  COMPosition ENSEmble mol2 FRACtional 0.78
</pre>
<a name="How_to_Define_Solutions" id="How_to_Define_Solutions"></a><h2> <span class="mw-headline">How to Define Solutions</span></h2>
<p>To include the files you should use the preprocessor command @
</p>
<pre>
  @ filename.sol
  @ filename.rlist
</pre>
<a name=".22sol.22_Files" id=".22sol.22_Files"></a><h3> <span class="mw-headline"> "sol" Files</span></h3>
<p>One copy of mol1 with known orientation and position (fractional coordinates)
</p>
<pre>   SOLUtion 6DIM ENSEmble mol1 EULEr 17 20 32 FRACtional 0.12 0.05 0.74
</pre>
<p>One copy of mol1 with known orientation only
</p>
<pre>   SOLUtion 3DIM ENSEmble mol1 EULEr 17 20 32
</pre>
<p>One copy of mol1 with known orientation and only the coordinates in 2 dimensions is known. The degenerate direction is defined as the direction perpendicular to the plane in which the position is given.
</p>
<pre>   SOLUtion 5DIM ENSEmble mol1 EULEr 17 20 32 DEGEnerate X FRACtional 0.05 0.74 
</pre>
<p>If the rotation function and translation function for mol1 were very clear, then there will only be one type of 6DIM solution for mol1. If the rotation and translation functions for mol2 were then not clear, there will be a series of possible 6DIM solutions for mol2.
</p>
<pre>
  SOLUtion SET 
  SOLUtion 6DIM ENSEmble mol1 EULEr 17 20 32 FRACtional 0.12 0.05 0.74
  SOLUtion 6DIM ENSEmble mol2 EULEr 5 183 230 FRACtional 0.71 0.54 0.81 
  SOLUtion SET
  SOLUtion 6DIM ENSEmble mol1 EULEr 17 20 32 FRACtional 0.12 0.05 0.74
  SOLUtion 6DIM ENSEmble mol2 EULEr 51 93 75 FRACtional 0.08 0.57 0.25 
</pre>
<a name=".22rlist.22_Files" id=".22rlist.22_Files"></a><h3> <span class="mw-headline">"rlist" Files</span></h3>
<p>THere are three trial orientations to search
</p>
<pre>
  SOLUtion TRIAl ENSEmble mol1 EULEr 17 20 32 SCORE 4.5
  SOLUtion TRIAl ENSEmble mol1 EULEr 67 65 51 SCORE 4.4
  SOLUtion TRIAl ENSEmble mol1 EULEr 67 112 81 SCORE 4.3
</pre>
<p>There are two possibilities for the position of the first molecule, and two orientations to search for the first and three for the second.
</p>
<pre>
  SOLUtion SET
  SOLUtion 6DIM ENSEmble mol1 EULEr 17 20 32 FRACtional 0.12 0.05 0.74
  SOLUtion TRIAl ENSEmble mol1 EULEr 44 20 32 SCORE 5.8
  SOLUtion TRIAl ENSEmble mol1 EULEr 67 65 51 SCORE 5.2
  SOLUtion SET
  SOLUtion 6DIM ENSEmble mol1 EULEr 17 20 32 FRACtional 0.13 0.55 0.76
  SOLUtion TRIAl ENSEmble mol1 EULEr 83 9 180 SCORE 6.3
  SOLUtion TRIAl ENSEmble mol1 EULEr 8 36 92 SCORE 4.2
  SOLUtion TRIAl ENSEmble mol1 EULEr 48 87 10 SCORE 4.0
</pre>
<p>If a degenerate translation function is performed, then a SOLUtion TRIAl line is produced with the degenerate translation information present, ready for performing the translation function on the third dimension.
</p>
<pre>
  SOLUtion TRIAl ENSEmble mol1 EULEr 17 20 32 DEGEnerate X FRACtional 0.05 0.74 
</pre>
<a name="Fixed_Partial_Structure" id="Fixed_Partial_Structure"></a><h3> <span class="mw-headline">Fixed Partial Structure</span></h3>
<p>If you have the coordinates of a partial solution with the pdb coordinates of the known structure in the correct orientation and position, then you can force Phaser to use these coordinates. Use this pdb file to define an ensemble (named "mol1" in this example). Then manually create a .sol file of the following form and include it in the Phaser command script with the @filename preprocessor command (or include it directly in the script)
</p>
<pre>
  SOLUtion SET 
  SOLUtion 6DIM ENSEmble mol1 EULEr 0 0 0 FRACtional 0 0 0
</pre>
<a name="Automated_Molecular_Replacement" id="Automated_Molecular_Replacement"></a><h2> <span class="mw-headline">Automated Molecular Replacement</span></h2>
<p>Example command script for finding BETA and BLIP. This is the minimum input, using all defaults (except the ROOT filename).
</p>
<pre>
  #''beta_blip_auto.com''
  phaser &lt;&lt; eof
  TITLe beta blip automatic
  MODE MR_AUTO
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  SEARch ENSEmble beta NUM 1
  SEARch ENSEmble blip NUM 1
  ROOT beta_blip_auto # not the default
  eof
</pre>
<a name="Non-Automated_Molecular_Replacement" id="Non-Automated_Molecular_Replacement"></a><h2> <span class="mw-headline">Non-Automated Molecular Replacement</span></h2>
<a name="Fast_Rotation_Function" id="Fast_Rotation_Function"></a><h3> <span class="mw-headline">Fast Rotation Function</span></h3>
<p>Example command script for fast rotation function to find the orientation of BETA.
</p>
<pre>  
  #beta_frf.com
  phaser &lt;&lt; eof
  TITLe beta FRF
  MODE MR_FRF
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  SEARCH ENSEmble beta
  ROOT beta_frf
  eof
</pre>
<p>Example command script for fast rotation function to find the orientation of BLIP knowing the position and orientation of BETA, with the position and orientation of BETA input from the command line.
</p>
<pre>
  #blip_frf_with_beta.com
  phaser &lt;&lt; eof
  TITLe blip FRF with beta rotation and translation
  MODE MR_FRF
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq #beta
  COMPosition PROTein SEQuence blip.seq #blip
  SEARch ENSEmble blip
  SOLUtion 6DIM ENSEmble beta EULEr 201 41 184 FRACtional -0.49408 -0.15571 -0.28148
  ROOT blip_frf_with_beta
  eof
</pre>
<p>Example command script for fast rotation function to find the orientation of BLIP knowing only the orientation of BETA, with the orientation of BETA input using the output solution file from the beta_frf.com job above.
</p>
<pre>
  #blip_frf_with_beta_rot.com
  phaser &lt;&lt; eof
  TITLe blip FRF with beta R
  MODE MR_FRF
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  SEARch ENSEmble blip
  @beta_frf.sol # solution file output by phaser
  ROOT blip_frf_with_beta_rot
  eof
</pre>
<a name="Brute_Rotation_Function" id="Brute_Rotation_Function"></a><h3> <span class="mw-headline">Brute Rotation Function</span></h3>
<p>Example command script for brute rotation function to find the orientation of BETA
</p>
<pre>
  #beta_brf.com
  phaser &lt;&lt; eof
  TITLe beta BRF
  MODE MR_BRF
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  SEARch ENSEmble beta
  ROOT beta_brf
  eof
</pre>
<p>Example command script for brute rotation function to find the optimal orientation of BETA in a restricted search range and on a fine grid around the position from the fast rotation search.
</p>
<pre>
  #beta_brf_around.com
  phaser &lt;&lt; eof
  TITLe beta BRF fine sampling
  MODE MR_BRF
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  SEARch ENSEmble beta
  ROTAte AROUnd EULEr 201 41 184 RANGE 10
  SAMPling ROTation 0.5
  XYZOut ON # not the default
  ROOT beta_brf_around
  eof
</pre>
<a name="Fast_Translation_Function" id="Fast_Translation_Function"></a><h3> <span class="mw-headline">Fast Translation Function</span></h3>
<p>Example command script for finding the position of BETA after the rotation function has been run and the results output to the file beta_frf.rlist
</p>
<pre>
  #beta_ftf.com
  phaser &lt;&lt; eof
  TITLe beta FTF
  MODE MR_FTF
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  @beta_frf.rlist
  ROOT beta_ftf
  eof 
</pre>
<p>Example command script for finding the position of BLIP after the rotation function has been run and the results output to the file blip_frf_with_beta.rlist, which has the SOLUtion 6DIM keyword input for BETA and the SOLUtion TRIAL keyword input for the orientations to try for BLIP with the translation function.
</p>
<pre>
  #blip_ftf_with_beta.com
  phaser &lt;&lt; eof
  TITLe beta FTF
  MODE MR_FTF
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  @blip_frf_with_beta.rlist
  ROOT blip_ftf_with_beta
  eof
</pre>
<a name="Brute_Translation_Function" id="Brute_Translation_Function"></a><h3> <span class="mw-headline">Brute Translation Function</span></h3>
<p>Example command script for brute Translation function to find the position of BETA after the rotation function has been run
</p>
<pre>  
  #beta_btf.com
  phaser &lt;&lt; eof
  TITLe beta BTF
  MODE MR_BTF
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  @beta_frf.rlist
  TRANslate AROUnd FRACtional POINt -0.49408 -0.15571 -0.28148 RANGe 5
  ROOT beta_btf
  eof 
</pre>
<p>Example command script for brute Translation function to find the position of BETA degenerate in X after the rotation function has been run
</p>
<pre>
  #beta_btf_degen_x.com
  phaser &lt;&lt; eof
  TITLe beta degenerate X
  MODE MR_BTF
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  @beta_frf.rlist
  TRANslate DEGEnerate X
  ROOT beta_btf_degen_x
  eof 
</pre>
<a name="Refinement_and_Phasing" id="Refinement_and_Phasing"></a><h3> <span class="mw-headline">Refinement and Phasing</span></h3>
<p>Example command script to refine a set of solutions
</p>
<pre>
  #beta_blip_rnp.com
  phaser &lt;&lt; eof
  TITLe beta blip rigid body refinement
  MODE MR_RNP
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  ROOT beta_blip_rnp # not the default
  HKLOut OFF # not the default
  XYZOut OFF # not the default
  @beta_blip_auto.sol
  eof 
</pre>
<a name="Log-Likelihood_Gain" id="Log-Likelihood_Gain"></a><h3> <span class="mw-headline">Log-Likelihood Gain</span></h3>
<p>Example command script to rescore the solutions using a different resolution range of data and a different spacegroup
</p>
<pre>
  #beta_blip_llg.com
  phaser &lt;&lt; eof
  TITLe beta blip solution 6A P3121
  MODE MR_LLG
  HKLIn beta_blip.mtz
  LABIn F=F SIGF = SIGF
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  ROOT beta_blip_llg # not the default
  RESOlution 6.0
  SPACegroup P 31 2 1
  @beta_blip_auto.sol
  eof 
</pre>
<a name="Packing" id="Packing"></a><h3> <span class="mw-headline">Packing</span></h3>
<p>Example command script for determining whether a set of molecular replacement solutions pack in the unit cell.
</p>
<pre>
  #beta_blip_pak.com
  phaser &lt;&lt; eof
  TITLe beta blip packing check
  MODE MR_PAK
  HKLIn beta_blip.mtz
  LABIn F=F SIGF=SIGF
  ENSEmble beta PDB beta.pdb IDENtity 100
  ENSEmble blip PDB blip.pdb IDENtity 100
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  ROOT beta_blip_pak # not the default
  @beta_blip_auto.sol
  eof 
</pre>
<a name="Automated_Experimental_Phasing" id="Automated_Experimental_Phasing"></a><h2> <span class="mw-headline">Automated Experimental Phasing</span></h2>
<p>Do SAD phasing of insulin. This is the minimum input, using all defaults (except the ROOT filename and specifying the wavelength explicitly).
</p>
<pre>
  #insulin_auto.com
  phaser &lt;&lt; eof
  MODE EP_AUTO
  TITLe sad phasing of insulin with intrinsic sulphurs
  HKLIn S-insulin.mtz
  COMPosition PROTein SEQ S-insulin.seq
  CRYStal insulin DATAset sad LABIn F+=F(+) SIG+=SIGF(+) F-=F(-) SIG-=SIGF(-)
  CRYStal insulin DATAset sad SCATtering CUKA # default: change if necessary
  LLGComplete CRYStal insulin COMPLETE ON SCATtering ELEMent S
  ATOM CRYStal insulin PDB S-insulin_hyss.pdb
  ROOT insulin_auto
  eof
</pre>
<a name="Anisotropy_Correction" id="Anisotropy_Correction"></a><h2> <span class="mw-headline">Anisotropy Correction</span></h2>
<p>Example command script to correct BETA-BLIP data for anisotropy
</p>
<pre> 
  #beta_blip_ano.com
  phaser &lt;&lt; eof
  MODE ANO
  TITLe beta blip data correction
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  ROOT beta_blip_ano # not the default
  eof 
</pre>
<a name="Cell_Content_Analysis" id="Cell_Content_Analysis"></a><h2> <span class="mw-headline">Cell Content Analysis</span></h2>
<p>Example script for cell content analysis for BETA-BLIP
</p>
<pre>
  #beta_blip_cca.com
  phaser &lt;&lt; eof
  TITLe BETA-BLIP cell content analysis
  MODE CCA
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  ROOT beta_blip_cca # not the default
  eof 
</pre>
<a name="Translational_NCS_Analysis" id="Translational_NCS_Analysis"></a><h2> <span class="mw-headline">Translational NCS Analysis</span></h2>
<p>Example script for translational NCS analysis for BETA-BLIP
</p>
<pre>
  #beta_blip_ncs.com
  phaser &lt;&lt; eof
  TITLe BETA-BLIP translational NCS analysis
  MODE NCS  
  HKLIn beta_blip.mtz
  LABIn F=Fobs SIGF=Sigma
  COMPosition PROTein SEQuence beta.seq NUM 1 #beta
  COMPosition PROTein SEQuence blip.seq NUM 1 #blip
  ROOT beta_blip_ncs # not the default
  eof 
</pre>
<a name="Normal_Mode_Analysis" id="Normal_Mode_Analysis"></a><h2> <span class="mw-headline">Normal Mode Analysis</span></h2>
<p>Do normal mode analysis, write out eigenfile and coordinates perturbed by default movements along mode 7 only. (If you only wanted to prepare the eigenfile but not coordinates, you could include the command "XYZOut OFF").
</p>
<pre>  
  #beta_nma.com
  phaser &lt;&lt; eof
  TITLe beta normal mode analysis
  MODE NMA
  ENSEmble beta PDB beta.pdb IDENtity 100
  ROOT beta_nma # not the default
  eof 
</pre>
<p>This example shows the use of several infrequently used options. Read in previous eigenfile and write out pdb files perturbed in 0.5 Ångstrom rms intervals in "forward" (positive dq values) direction only along modes 7 and 10 (and combinations of 7 and 10), up to a maximum rms shift of 1.2Å. Normally you would want to perturb the structure in both directions, and modes 8 and 9 are more likely to be of interest than mode 10, but something like this might be useful as part of a larger, more exhaustive, exploration.
</p>
<pre>
  #beta_nma_pdb.com
  phaser &lt;&lt; eof
  TITLe beta normal mode analysis pdb file generation
  MODE NMA
  ENSEmble beta PDB beta.pdb IDENtity 100
  ROOT beta_nma_pdb # not the default
  EIGEn READ beta_nma.mat
  NMAPdb MODE 7 MODE 10 
  NMAPdb RMS STEP 0.5 
  NMAPdb RMS MAXRMS 1.2 
  NMAPDB RMS DIRECTION FORWARD
  eof
</pre>

<!-- 
NewPP limit report
Preprocessor node count: 150/1000000
Post-expand include size: 0/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->
<div class="printfooter">
</div>
	    	    <!-- end content -->
	    <div class="visualClear"></div>
	  </div>
	</div>
      </div>
      <div id="column-one">
	<div class="portlet" id="p-logo">
            <a style="background-image: url(../../../../misc/PhaserwikiLogo.png);"
	    href="http://www.phaser.cimr.cam.ac.uk/index.php/Phaser_Crystallographic_Software"
	    title="Phaser Crystallographic Software"></a>
	</div>
	<script type="text/javascript"> if (window.isMSIE55) fixalpha(); </script>
		<div class='portlet' id='p-site'>
	  <h5>site</h5>
	  <div class='pBody'>
	    <ul>
	    	      <li id="n-Home"><a href="../../../../index.html">Home</a></li>
	     	      <li id="n-News"><a href="../../../../articles/n/e/w/News.html">News</a></li>
	     	      <li id="n-Events"><a href="../../../../articles/e/v/e/Events.html">Events</a></li>
	     	      <li id="n-Downloads"><a href="../../../../articles/d/o/w/Downloads.html">Downloads</a></li>
	     	    </ul>
	  </div>
	</div>
		<div class='portlet' id='p-documentation'>
	  <h5>documentation</h5>
	  <div class='pBody'>
	    <ul>
	    	      <li id="n-Manuals"><a href="../../../../articles/m/a/n/Manuals.html">Manuals</a></li>
	     	      <li id="n-MR-Phasing"><a href="../../../../articles/m/o/l/Molecular_Replacement_a23a.html">MR Phasing</a></li>
	     	      <li id="n-SAD-Phasing"><a href="../../../../articles/e/x/p/Experimental_Phasing_5400.html">SAD Phasing</a></li>
	     	      <li id="n-Keyword-Examples"><a href="../../../../articles/m/r/_/MR_using_keyword_input_9c88.html">Keyword Examples</a></li>
	     	      <li id="n-Python-Examples"><a href="../../../../articles/p/y/t/Python_Example_Scripts_7723.html">Python Examples</a></li>
	     	      <li id="n-Tutorials"><a href="../../../../articles/t/u/t/Tutorials.html">Tutorials</a></li>
	     	      <li id="n-FAQ"><a href="../../../../articles/f/a/q/FAQ_1fe9.html">FAQ</a></li>
	     	      <li id="n-Top-Ten-Tips"><a href="../../../../articles/t/o/p/Top_Ten_Tips_3fcd.html">Top Ten Tips</a></li>
	     	      <li id="n-Bugs"><a href="../../../../articles/b/u/g/Bugs.html">Bugs</a></li>
	     	      <li id="n-Publications"><a href="../../../../articles/p/u/b/Publications.html">Publications</a></li>
	     	      <li id="n-View-SVN"><a href="../../../../articles/s/v/n/SVN_Repository_93a5.html">View SVN</a></li>
	     	    </ul>
	  </div>
	</div>
		<div class='portlet' id='p-other'>
	  <h5>other</h5>
	  <div class='pBody'>
	    <ul>
	    	      <li id="n-Contact"><a href="../../../../articles/c/o/n/Contact.html">Contact</a></li>
	     	      <li id="n-Developers"><a href="../../../../articles/d/e/v/Developers.html">Developers</a></li>
	     	      <li id="n-Licences"><a href="../../../../articles/l/i/c/Licences.html">Licences</a></li>
	     	      <li id="n-External-Links"><a href="../../../../articles/e/x/t/External_Links_ff0e.html">External Links</a></li>
	     	      <li id="n-Help"><a href="../../../../articles/c/o/n/Help%7EContents_22de.html">Help</a></li>
	     	    </ul>
	  </div>
	</div>
	<!--
	<div id="p-search" class="portlet">
	  <h5><label for="searchInput">Search</label></h5>
	  <div id="searchBody" class="pBody">
	    <form action="javascript:goToStatic(3)" id="searchform"><div>
	      <input id="searchInput" name="search" type="text"
	        accesskey="f" value="" />
	      <input type='submit' name="go" class="searchButton" id="searchGoButton"
	        value="Go" />
	    </div></form>
	  </div>
	</div>
-->
	      </div><!-- end of the left (by default at least) column -->
      <div class="visualClear"></div>
      <div id="footer">
<!--
    <div id="f-poweredbyico"><a href="http://www.mediawiki.org/"><img src="../../../../skins/common/images/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" /></a></div>	-->
<!-- date DATE Date -->
	<ul id="f-list">
	   Static <a href="http://www.phaser.cimr.cam.ac.uk/index.php/Phaser_Crystallographic_Software">Phaserwiki</a> 
Thu Jul 25 23:20:16 BST 2013
<!--
	            </li>
	  	  <li id="f-credits">This page was last modified 21:34, 1 August 2012 by <a href="../../../../articles/a/i/r/User%7EAirlie_8634.html" title="User:Airlie">Airlie McCoy</a>. Based on work by <a href="../../../../articles/r/a/n/User%7ERandy_8422.html" title="User:Randy">Randy Read</a>&#32;and&#32;<a href="../../../../articles/w/i/k/User%7EWikiSysop_5eaf.html" title="User:WikiSysop">WikiSysop</a>.</li>	  	  <li id="f-about"><a href="../../../../articles/a/b/o/Phaserwiki%7EAbout_932f.html" title="Phaserwiki:About">About Phaserwiki</a></li>	  	  -->
	</ul>
      </div>
    </div>
  </body>
</html>
