<!DOCTYPE html
                     PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
                     "DTD/xhtml1-transitional.dtd">
<html>
<head>
<title>CCP4 Interface: Coordinate Utilities Module</title>
<link rel="stylesheet" type="text/css" href="../ccp4idocs.css" title="CCP4i" />
<link rev="made" href="mailto:mgwt@ysbl.york.ac.uk" />
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<meta name="AUTHOR" content="Liz and Maria" />
<meta name="CHANGEDBY" content="Maria" />
<meta name="CHANGED" content="20000524" />
<meta name="GENERATOR" content="Mozilla/3.0Gold (X11; I; IRIX 6.3 IP32) [Netscape]" />
</head>
<body>

<table border="0">
<tr><td rowspan="2"><img src="../images/weblogo175.gif" alt="CCP4 web logo" /></td>
<td class="logohead">CCP4i: Graphical User Interface</td></tr>
<tr><td class="logotitle">Coordinate Utilities Module</td></tr>
</table>
<br />
<table border="0">
<tr><td>
<a href="viewer_utils.html" target="_self">
<img border="0" class="navbutton" src="../images/3Dnexttr.gif" alt="next button" /></a>
<a href="mtz_utils.html" target="_self">
<img border="0" class="navbutton" src="../images/3Dprevtr.gif" alt="previous button" /></a>
<a href="../index.html" target="_top">
<img border="0" class="navbutton" src="../images/3Dtoptr.gif" alt="top button" /></a>
</td></tr>
</table>

<dl>
<dt><font color="#cc3333"><strong>Tasks in this module:</strong></font></dt>
</dl>

<dl>
<dt><a href="#matthews" target="_self">Cell Content Analysis</a></dt>
<dd><a href="#matthews_layout" target="_self">Matthews - Task Window Layout</a></dd>
</dl>
<dl>
<dt><a href="#coordconv" target="_self">Convert Coordinate Formats</a></dt>
</dl>
<dl>
<dt><a href="#pdbset" target="_self">Edit PDB File</a></dt>
<dd><a href="#pdbcur_layout" target="_self">PDBCUR - Task Window Layout</a></dd>
<dd><a href="#pdbset_layout" target="_self">PDBSET - Task Window Layout</a></dd>
</dl>
<dl>
<dt><a href="#superpose" target="_self">Superpose Molecules</a></dt>
<dd><a href="#superpose_layout" target="_self">Superpose - Task Window Layout</a></dd>
</dl>
<dl>
<dt><a href="#edit_tls" target="_self">Create/Edit TLS File</a></dt>
</dl>
<dl>
<dt><a href="#get_prot" target="_self">Import/Edit Protein Sequence</a></dt>
</dl>
<dl>
<dt><a href="#clustalw" target="_self">ClustalW Interface</a></dt>
</dl>

<p>The layout of each task window, <em>i.e.</em> the number of folders present,
and whether these folders are open or closed by default, depends on the choices
made in the <strong>Protocol</strong> folder of the task (see
<a href="../general/intro.html#task_window_format" target="_self">Introduction</a>). Although
certain folders are closed by default, there are specific reasons why you should
or may want to look at them. These reasons are described in the Task Window
Layout sections below.</p>

<h3><a name="matthews"></a><a name="matthews2"></a>Cell Content Analysis</h3>

<p>Both '<tt>dm</tt>' and Solomon require the input of a number for the fraction of
solvent in the unit cell. The program MATTHEWS_COEF, through the Cell Content Analysis task,
can provide this number. It requires an MTZ file, and a fairly accurate estimate of the
molecular weight of the protein. The latter can be given directly (obtained from the 
program <a href="../../../../html/rwcontents.html" target="_PROGDOC">RWCONTENTS</a>, for
example), or the task will estimate it from a supplied number of residues (per molecule). </p>

<h4><a name="matthews_layout"></a>Matthews - Task Window Layout</h4>

<p>The solvent content information will appear in the 'solvent analysis' field upon
clicking <tt>Run Now</tt>. The Interface displays a table with values of the percentage
of solvent in the unit cell, as well as the corresponding Matthews coefficient, for a
range of numbers of molecules in the asymmetric unit, from which the user reads the
correct one.</p>

<p>Raw scores are also given for each line based on the paper of 
<a href="../../../../html/matthews_coef.html#reference2" target="_PROGDOC">Kantardjieff and Rupp</a>.
<em>score(reso)</em> is the score using the high resolution limit obtained from the
MTZ file, and <em>score(tot)</em> is the score across all resolution ranges.</p>

<p>See program documentation:
<a href="../../../../html/matthews_coef.html" target="_PROGDOC">MATTHEWS_COEF</a>.</p>

<h3><a name="coordconv"></a><a name="coordconv2"></a>Convert Coordinate Formats</h3>

<p>The program COORDCONV is used to convert coordinate files from various formats
(including SHELX, CSD and XPLOR-style formats) to PDB, CCP4 and XPLOR formats. The new
CCP4 Heavy Atom format is included, too. Take care in using the correct orthogonalisation
code.</p>

<p>See program documentation:
<a href="../../../../html/coordconv.html" target="_PROGDOC">COORDCONV</a>.</p>

<h3><a name="pdbset"></a><a name="pdbset2"></a>Edit PDB File</h3>

<p>The task runs with a choice of PDBCUR or PDBSET as the underlying program. In the
Protocol folder, the user should select the program and then select from the list
of options. PDBCUR and PDBSET provide many manipulations of coordinate files: some are
common to both programs, some are unique to one program, and not all are interfaced
in ccp4i.</p>

<p>The program PDBCUR typically takes coordinate files in PDB format, and the interface
is written assuming this. However, the task will in fact accept coordinate files in mmCIF 
or MMDB binary format.</p>

<p>The program PDBSET is used for various manipulations on coordinate files.
Apart from PDB format, the input file may be <a href="../../../../html/pdbset.html#xplor"
target="_PROGDOC">XPLOR</a> format, in which case certain operations are performed
automatically.</p>

<h4><a name="pdbcur_layout"></a>PDBCUR - Task Window Layout</h4>

<p>The options "remove hydrogen atoms" and "remove alternative conformations" run
without any further parameters. The option "remove atoms with low/zero occupancy"
accepts an occupancy cut-off. Leave it at the default of 0.0 if you just want to
remove zero occupancy atoms.</p>

<h4><a name="pdbset_layout"></a>PDBSET - Task Window Layout</h4>

<p>Not all keywords available to the PDBSET program, are interfaced in CCP4i. For instance,
it is not possible to change the orthogonalisation code (which <em>can</em> be changed
with the task <a href="#coordconv" target="_self">Convert Coordinate Formats</a>).</p>

<p>The type of input file (PDB or XPLOR) is set in the Protocol folder. If this file
contains appropriate symmetry information, this is used in the manipulations. The Interface
has the option to use the symmetry information from an MTZ file (in the Files folder),
which is not an option in the PDBSET program itself. Any symmetry information already
present in coordinate and/or reflection files, can then be reset in the
'Required Parameters' folder, through <tt>Reset cell parameters</tt> (check the little
square box).</p>

<p>See program documentation:
<a href="../../../../html/pdbcur.html" target="_PROGDOC">PDBCUR</a>.
<a href="../../../../html/pdbset.html" target="_PROGDOC">PDBSET</a>.</p>

<h3><a name="superpose"></a><a name="superpose1"></a>Superpose Molecules</h3>

<p>There are three approaches to superposing molecules available:</p>
<ol type="a">
<li>Using the Secondary Structure Matching (SSM) algorithm (using the SUPERPOSE program)</li>
<li>Superposing atoms and/or residues specified by the user (using the LSQKAB program)</li>
<li>Finding the best fit of secondary structure elements (using the TOP program)
<p>There is a fourth use of this task:</p></li>
<li>Transformation of one molecule, if the translation and rotation to be applied are
known (using the LSQKAB program)</li>
</ol>

<p>The SSM algorithm does not require prior knowledge of which residues are equivalent
in the two structures. Atom selection can be used to superpose subsets of the two structures.
By default, the task outputs a PDB file containing the first structure rotated and
translated according to the determined superposition. The whole structure is transformed,
even if the superposition was only calculated on a subset. Optionally, the transformed
structure can be merged with the fixed structure (using the program
<a href="../../../../html/pdb_merge.html" target="_PROGDOC">PDB_MERGE</a>). (See
also the
SUPERPOSE <a href="../../../../html/superpose.html" target="_PROGDOC">program documentation</a>.)</p>

<p>The TOP program also does not require prior knowledge of which residues are equivalent and can 
usually superpose any structures which have the same basic fold. It reports the sequence alignment of
the residues which it finds to be equivalent and it may report multiple solutions when there is
no one clear solution. Coordinates for the best fit solution are output.</p>

<p>The TOP program has functionality to search databases for similar structures which is not
currently interfaced by CCP4i (see <a href="../../../../html/topp.html" target="_PROGDOC">program
documentation</a>).</p>

<h4><a name="superpose_layout"></a>Superpose - Task Window Layout</h4>

<p>In the Protocol folder, one of the
<a href="#superpose" target="_self">four above-mentioned approaches</a> is chosen.
In the Files folder, enter the coordinate filenames for one molecule which remains fixed
and one which is moved, and one for the coordinates output to another PDB file. The number
of input and output files to be specified, depends on the Protocol option chosen.</p>

<p>If SSM (option a) is chosen, then there is a folder for Atom Selection. The program uses 
the atom selection syntax described in the 
<a href="../../../../html/pdbcur.html#atom_selection" target="_PROGDOC">PDBCUR</a>
documentation. This can be added directly, or a
template created with the boxes provided and then edited if necessary.</p>

<p>
If the LSQKAB program is used for superposition of specified atoms/residues (option b),
there are also options to
<ul>
<li>output the RMS differences in coordinates of the superposed residues to a file called
<em>..rmstab.graph</em> which, when selected from the <tt>View Files from Job</tt> utility,
will be displayed by <a href="../general/loggraph.html" target="_self">Loggraph</a>.</li>
<li>output all differences between atom pairs to a file assigned to DELTAS.</li>
</ul>
LSQKAB reports the translation and rotation applied in the log file.</p>

<p>For Protocol option d, the input in the 'Define Transformation' folder needs to be in
the following format:</p>
<table border="0">
<tr><td>Rotation</td><td>Eulerian angles</td></tr>
<tr><td>Translation</td><td>&#197;ngstr&#246;m along cell edges</td></tr>
</table>
<p>The input coordinates are rotated first, and then translated.</p>

<p>See program documentation:
<a href="../../../../html/superpose.html" target="_PROGDOC">SUPERPOSE</a> (SSM),
<a href="../../../../html/lsqkab.html" target="_PROGDOC">LSQKAB</a>,
<a href="../../../../html/topp.html" target="_PROGDOC">TOP</a>.<br />
For more information on TOP and related programs, visit
<a href="http://gamma.mbb.ki.se/~guoguang/" target="_EXTERNAL">Guoguang Lu</a>.</p>

<h3><a name="edit_tls"></a><a name="edit_tls1"></a>Create/Edit TLS File</h3>

<p>See description in <a href="refinement.html#edit_tls">Refinement</a>
module.</p>

<h3><a name="get_prot"></a><a name="get_prot1"></a>Import/Edit Protein Sequence</h3>

<p>The Import/Edit Protein Sequence Task is used for importing and manipulating protein 
sequences obtained from the SwissProt database at the EBI (alternative database sources
can be specified in <a href="../general/configure.html#sequence_database">Configure Interface</a>). 
Enter the SwissProt code of the protein sequence to be
viewed and a summary view of the protein sequence appears in the Current Sequence window, with
a full detailed view in the View Full Sequence Entry window. The Editing window contains editing
tools to manipulate the protein sequence in the Current Sequence window. Enter the search string,
click the <tt>Search For...</tt> button and if the search string is present in the protein
sequence, it is highlighted. This section of the protein sequence can then be deleted, mutated
into another specified sequence or amino acids can be inserted after this section. The changed
sequence can then be saved in a file of your specification for future viewing and editing.</p>

<p>This program is the first CCP4i program to connect to a database over the internet in order
to download a requested file. If you use a proxy server, please remember to set this 
in <a href="../general/configure.html#proxy">Configure Interface</a> before running CCP4i.</p>

<p>See full program documentation:
<a href="../../../../html/get_prot.html" target="_PROGDOC">GET_PROT</a></p>

<h3><a name="clustalw"></a><a name="clustalw1"></a>ClustalW Interface</h3>

<p>Interface to run the ClustalW program. (See the full Program Documentation
for the <a href="../../../../html/clustalw.html">ClustalW Interface</a> for more
information).<em>N.B. ClustalW is not distributed as part of CCP4 and needs to
be obtained separately</em>.</p>

<table border="0">
<tr><td>
<a href="programlist.html" target="_self">
<img border="0" class="navbutton" src="../images/3Dnexttr.gif" alt="next button" /></a>
<a href="mtz_utils.html" target="_self">
<img border="0" class="navbutton" src="../images/3Dprevtr.gif" alt="previous button" /></a>
<a href="../index.html" target="_top">
<img border="0" class="navbutton" src="../images/3Dtoptr.gif" alt="top button" /></a>
</td>
<td align="right"><a href="http://validator.w3.org/" target="_EXTERNAL">Valid XHTML 1.0!</a>
<a href="http://jigsaw.w3.org/css-validator/" target="_EXTERNAL">Valid CSS!</a></td></tr>
</table>

</body>
</html>
