<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en" dir="ltr">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
	<!-- headlinks removed -->
	<link rel="shortcut icon" href="../../../../misc/favicon.ico"/>
    <title>Molecular Replacement - Phaserwiki</title>
    <style type="text/css">/*<![CDATA[*/ @import "../../../../skins/offline/main.css"; /*]]>*/</style>
    <link rel="stylesheet" type="text/css" media="print" href="../../../../skins/common/commonPrint.css" />
    <!--[if lt IE 5.5000]><style type="text/css">@import "../../../../skins/monobook/IE50Fixes.css";</style><![endif]-->
    <!--[if IE 5.5000]><style type="text/css">@import "../../../../skins/monobook/IE55Fixes.css";</style><![endif]-->
    <!--[if IE 6]><style type="text/css">@import "../../../../skins/monobook/IE60Fixes.css";</style><![endif]-->
    <!--[if IE]><script type="text/javascript" src="../../../../skins/common/IEFixes.js"></script>
    <meta http-equiv="imagetoolbar" content="no" /><![endif]-->
    <script type="text/javascript" src="../../../../skins/common/wikibits.js"></script>
    <script type="text/javascript" src="../../../../skins/offline/md5.js"></script>
    <script type="text/javascript" src="../../../../skins/offline/utf8.js"></script>
    <script type="text/javascript" src="../../../../skins/offline/lookup.js"></script>
    <script type="text/javascript" src="../../../../raw/gen.js"></script>                  </head>
  <body
    class="ns-0 ns-subject page-Molecular_Replacement">
    <div id="globalWrapper">
      <div id="column-content">
	<div id="content">
	  <a name="top" id="contentTop"></a>
	        <h1 class="firstHeading">Molecular Replacement</h1>
	  <div id="bodyContent">
	    <h3 id="siteSub">From Phaserwiki</h3>
	    <div id="contentSub"></div>
	    	    	    <!-- start content -->
	    <div style="margin-left: 25px; float: right;"><table id="toc" class="toc" summary="Contents"><tr><td><div id="toctitle"><h2>Contents</h2></div>
<ul>
<li class="toclevel-1"><a href="#Automated_Molecular_Replacement"><span class="tocnumber">1</span> <span class="toctext">Automated Molecular Replacement</span></a></li>
<li class="toclevel-1"><a href="#Should_Phaser_Solve_It.3F"><span class="tocnumber">2</span> <span class="toctext">Should Phaser Solve It?</span></a></li>
<li class="toclevel-1"><a href="#What_Resolution_of_Data_Should_be_Used.3F"><span class="tocnumber">3</span> <span class="toctext">What Resolution of Data Should be Used?</span></a></li>
<li class="toclevel-1"><a href="#Has_Phaser_Solved_It.3F"><span class="tocnumber">4</span> <span class="toctext">Has Phaser Solved It?</span></a></li>
<li class="toclevel-1"><a href="#Annotation"><span class="tocnumber">5</span> <span class="toctext">Annotation</span></a></li>
<li class="toclevel-1"><a href="#What_to_do_in_Difficult_Cases"><span class="tocnumber">6</span> <span class="toctext">What to do in Difficult Cases</span></a></li>
<li class="toclevel-1"><a href="#How_to_Define_Data"><span class="tocnumber">7</span> <span class="toctext">How to Define Data</span></a></li>
<li class="toclevel-1"><a href="#How_to_Define_Models"><span class="tocnumber">8</span> <span class="toctext">How to Define Models</span></a>
<ul>
<li class="toclevel-2"><a href="#Building_an_Ensemble_from_Coordinates"><span class="tocnumber">8.1</span> <span class="toctext">Building an Ensemble from Coordinates</span></a>
<ul>
<li class="toclevel-3"><a href="#Coordinate_Editing"><span class="tocnumber">8.1.1</span> <span class="toctext">Coordinate Editing</span></a>
<ul>
<li class="toclevel-4"><a href="#HETATM.2FLIGANDS"><span class="tocnumber">8.1.1.1</span> <span class="toctext">HETATM/LIGANDS</span></a></li>
<li class="toclevel-4"><a href="#WATER"><span class="tocnumber">8.1.1.2</span> <span class="toctext">WATER</span></a></li>
</ul>
</li>
</ul>
</li>
<li class="toclevel-2"><a href="#Building_an_Ensemble_from_Electron_Density"><span class="tocnumber">8.2</span> <span class="toctext">Building an Ensemble from Electron Density</span></a></li>
</ul>
</li>
<li class="toclevel-1"><a href="#How_to_Define_Composition"><span class="tocnumber">9</span> <span class="toctext">How to Define Composition</span></a>
<ul>
<li class="toclevel-2"><a href="#Default_Composition"><span class="tocnumber">9.1</span> <span class="toctext">Default Composition</span></a></li>
<li class="toclevel-2"><a href="#Composition_by_Solvent_Content"><span class="tocnumber">9.2</span> <span class="toctext">Composition by Solvent Content</span></a></li>
<li class="toclevel-2"><a href="#Composition_by_Number_of_Residues_in_ASU"><span class="tocnumber">9.3</span> <span class="toctext">Composition by Number of Residues in ASU</span></a></li>
<li class="toclevel-2"><a href="#Composition_by_Molecular_Weight"><span class="tocnumber">9.4</span> <span class="toctext">Composition by Molecular Weight</span></a></li>
<li class="toclevel-2"><a href="#Composition_by_Sequence"><span class="tocnumber">9.5</span> <span class="toctext">Composition by Sequence</span></a></li>
<li class="toclevel-2"><a href="#Composition_by_Percentage_Scattering"><span class="tocnumber">9.6</span> <span class="toctext">Composition by Percentage Scattering</span></a></li>
</ul>
</li>
<li class="toclevel-1"><a href="#How_to_Define_Solutions"><span class="tocnumber">10</span> <span class="toctext">How to Define Solutions</span></a>
<ul>
<li class="toclevel-2"><a href="#.22sol.22_Files"><span class="tocnumber">10.1</span> <span class="toctext">"sol" Files</span></a></li>
<li class="toclevel-2"><a href="#.22rlist.22_Files"><span class="tocnumber">10.2</span> <span class="toctext">"rlist" Files</span></a></li>
<li class="toclevel-2"><a href="#Fixed_partial_structure"><span class="tocnumber">10.3</span> <span class="toctext">Fixed partial structure</span></a></li>
</ul>
</li>
<li class="toclevel-1"><a href="#How_to_Select_Peaks"><span class="tocnumber">11</span> <span class="toctext">How to Select Peaks</span></a></li>
<li class="toclevel-1"><a href="#How_to_Control_Output"><span class="tocnumber">12</span> <span class="toctext">How to Control Output</span></a></li>
</ul>
</td></tr></table><script type="text/javascript"> if (window.showTocToggle) { var tocShowText = "show"; var tocHideText = "hide"; showTocToggle(); } </script>
</div>
<p>Phaser should be able to solve most structures with the Automated Molecular Replacement mode, and this is the first mode that you should try.  Give Phaser your data (<a href="#How_to_Define_Data" title="">How to Define Data</a>) and your models (<a href="#How_to_Define_Models" title="">How to Define Models</a>), tell Phaser what to search for, and a list of possible spacegroups (in the same point group).
</p><p>If this doesn't work (see <a href="#Has_Phaser_Solved_It.3F" title=""> Has Phaser Solved It?</a>), you can try selecting peaks of lower significance in the rotation function in case the real orientation was not within the selection criteria. By default peaks above 75% of the top peak are selected (see <a href="#How_to_Select_Peaks" title=""> How to Select Peaks</a>). See <a href="#What_to_do_in_Difficult_Cases" title=""> What to do in Difficult Cases</a> for more hints and tips. If the automated molecular replacement mode doesn't work even with non-default input you need to run the modes of Phaser separately. The possibilities are endless - you can even try exhaustive searches (translations of all orientations) if you want - but experience has shown that most structures that can be solved by Phaser can be solved by relatively simple strategies.
</p>
<a name="Automated_Molecular_Replacement" id="Automated_Molecular_Replacement"></a><h2> <span class="mw-headline">Automated Molecular Replacement</span></h2>
<p>Automated Molecular Replacement combines the anisotropy correction, likelihood enhanced fast rotation function, likelihood enhanced fast translation function, packing and refinement modes for multiple search models and a set of possible spacegroups to automatically solve a structure by molecular replacement. Top solutions are output to the files FILEROOT.sol, FILEROOT.#.mtz and FILEROOT.#.pdb (where "#" refers to the sorted solution number, 1 being the best, and only 1 is output by default). Many structures can be solved by running an automated molecular replacement search with defaults, giving the ensembles that you expect to be easiest to find first.
</p>
<div class="thumb tnone"><div class="thumbinner" style="width:102px;"><a href="../../../../articles/p/h/a/File%7EPhaser_MR_auto.gif_c365.html" class="image" title="Flow Diagram for Automated MR"><img alt="" src="../../../../images/thumb/7/70/Phaser_MR_auto.gif/100px-Phaser_MR_auto.gif" width="100" height="67" border="0" class="thumbimage" /></a>  <div class="thumbcaption"><div class="magnify"><a href="../../../../articles/p/h/a/File%7EPhaser_MR_auto.gif_c365.html" class="internal" title="Enlarge"><img src="../../../../skins/common/images/magnify-clip.png" width="15" height="11" alt="" /></a></div>Flow Diagram for Automated MR</div></div></div>
<a name="Should_Phaser_Solve_It.3F" id="Should_Phaser_Solve_It.3F"></a><h2> <span class="mw-headline">Should Phaser Solve It?</span></h2>
<p>The difficulty of a molecular replacement problem depends primarily on two major factors: how well the model will be able to explain the diffraction data (which depends both on the accuracy of the model and on its completeness), and how many reflections can be explained, at least in part.  Each reflection provides a piece of information that helps to identify correct MR solutions.
</p><p>It is possible to make a reasonable prediction of whether or not a solution will be found.  If the quality of the model (its accuracy and completeness) can be estimated, then the expected contribution of each reflection to the total LLG can also be estimated.  From a large battery of tests, we know that an LLG of 40 or greater usually indicates a correct solution (at least in the absence of complicating factors such as translational non-crystallographic symmetry, tNCS).  Building on this understanding, if it is estimated that the LLG will be 60 or less, then Phaser will assume that the problem is a difficult one, and will implement search procedures optimised for difficult problems.
</p>
<a name="What_Resolution_of_Data_Should_be_Used.3F" id="What_Resolution_of_Data_Should_be_Used.3F"></a><h2> <span class="mw-headline">What Resolution of Data Should be Used?</span></h2>
<p>The signal for a molecular replacement solution should be very clear if the expected value of the LLG is much higher than the minimum required to be fairly certain of a solution.  Currently Phaser aims for a minimum LLG of 120 and, if it is possible to achieve an even higher value, given the quality of the model and the quantity of diffraction data, then the resolution for the initial search is limited to the value required to achieve an expected LLG of 120.  Data to the full resolution are still used for a final rigid-body refinement, or in a second pass if a clear solution is not found in the first attempt.
</p><p>However, if the model is expected to have a large RMS error (based usually on the correlation between sequence identity and RMS error), then data to high resolution will not contribute any significant signal.  Regardless of the expected LLG at the highest resolution limit, the resolution used is limited to 1.8 times the estimated RMS error of the model, because this resolution limit gives about 99% of the LLG that could be achieved.
</p><p>Because Phaser implements strategies designed to solve structures with as much confidence as possible, as efficiently as possible, it is best to leave the choice of resolution to Phaser, at least in the first instance.
</p>
<a name="Has_Phaser_Solved_It.3F" id="Has_Phaser_Solved_It.3F"></a><h2> <span class="mw-headline">Has Phaser Solved It?</span></h2>
<table class="wikitable" style="margin-left: 30px" align="right">

<tr>
<th> TF Z-score </th><th> Have I solved it?
</th></tr>
<tr>
<td> less than 5 </td><td> no
</td></tr>
<tr>
<td> 5 - 6 </td><td> unlikely
</td></tr>
<tr>
<td> 6 - 7 </td><td> possibly
</td></tr>
<tr>
<td> 7 - 8  </td><td> probably
</td></tr>
<tr>
<td> more than 8* </td><td>definitely
</td></tr>
<tr>
<td colspan="2" style="text-align: center;"> *<i>6 for 1st model in monoclinic space groups</i>
</td></tr></table>
<p>Ideally, a unique solution with a strong signal will be found at the end of the search.  If you are searching for multiple components, then ideally the search for each component will also give a strong signal.  However if the signal-to-noise of your search is low, there will be noise peaks and multiple ambiguous solutions.    Signal-to-noise is judged using the <b>Z-score</b>, which is computed by comparing the LLG values from the rotation or translation search with LLG values for a set of random rotations or translations. The mean and the RMS deviation from the mean are computed from the random set, then the Z-score for a search peak is defined as its LLG minus the mean, all divided by the RMS deviation, <i>i.e. </i> <b>the number of  standard deviations above (or below) the mean. </b>
</p><p>For a rotation function, the correct solution may be in the list with a Z-score under 4, and will not be found until a translation function is performed and picks out the correct solution.
</p><p>For a translation function the correct solution will generally have a Z-score (number of standard deviations above the mean value) over 5 and be well separated from the rest of the solutions.  Of course, there will always be exceptions! The table gives a very rough guide to interpreting TF Z-scores.  This table will be updated, as we learn more from systematic molecular replacement trials.
</p><p>When you are searching for multiple components, the signal may be low for the first few components but, as the model becomes more complete, the signal should become stronger.  Finding a clear solution for a new component is a good sign that the partial solution to which that component was added was indeed correct.
</p><p>You should always at least glance through the summary of the logfile. One thing to look for, in particular, is whether any translation solutions with a high Z-score have been rejected by the packing step. By default up to 5 percent of marker atoms (C-alpha atoms for protein) are allowed to be involved in clashes. A solution with more clashes may still be correct, and the clashes may arise only because of differences in small surface loops.  If this happens, repeat the run allowing a suitable number of clashes.  Note that, unless there is specific evidence in the logfile that a high TF-function Z-score solution is being rejected with a few clashes, it is much better to edit the model to remove the loops than to increase the number of allowed clashes. Packing criteria are a very powerful constraint on the translation function, and increasing the number of allowed clashes beyond the default will increase the search time enormously without the possibility of generating any correct solutions that would not have otherwise been found.
</p><p>Note that, by default, Phaser will produce a single PDB file corresponding to the top solution found (if any), so finding a single PDB file in your output directory is not an indication that the search succeeded!  You have to look, at least, at the summary of the logfile, or at the list of possible solutions in the .sol file that is produced if you run Phaser from ccp4i or command-line scripts.
</p>
<a name="Annotation" id="Annotation"></a><h2> <span class="mw-headline">Annotation</span></h2>
<p>A highly compact summary of the history of a solution is given in the annotation of a solution in the .sol file.  This is a good place to start your analysis of the output. The annotation gives the Z-score of the solution at each rotation and translation function, the number of clashes in the packing, and the refined LLG.
</p>
<table class="wikitable" style="margin-left: 30px">

<tr>
<th>   Annotation   </th><th> Meaning
</th></tr>
<tr>
<td> RFZ= </td><td> Rotation Function Z-score
</td></tr>
<tr>
<td> TFZ= </td><td> Translation Function Z-score
</td></tr>
<tr>
<td> PAK= </td><td> Number of packing clashes
</td></tr>
<tr>
<td> LLG= </td><td> LLG after refinement (resolution may change)
</td></tr>
<tr>
<td> TFZ== </td><td> Translation Function Z-score equivalent, only calculated for the top solution after refinement (or for the number of top files specified by TOPFILES)
</td></tr>
<tr>
<td> RF++  </td><td> Rotation angle from previous Rotation Function has been used in the addition of next solution
</td></tr>
<tr>
<td> RF*0  </td><td> Rotation angle 000 identified by low R-factor of input model
</td></tr>
<tr>
<td> TFZ=* </td><td> First molecule in P1 (arbitrary origin, no Translation Function required)
</td></tr>
<tr>
<td> TF*0  </td><td> Translation vector 000 identified by low R-factor of input model
</td></tr>
<tr>
<td> (&amp;&nbsp;... &amp; ...) </td><td> Set of TFZ PAK and LLG values for placements that were amalgamated (more than one placement from a single Translation Function)
</td></tr>
<tr>
<td> LLG+=(...&nbsp;&amp;&nbsp;...)&nbsp;</td><td> Set of LLG values calculated during amalgamation, which will always be increasing in value
</td></tr>
<tr>
<td> +TNCS </td><td> Components added by Translational NCS relation
</td></tr></table>
<p>Two versions of TFZ (the translation function Z-score) now appear for each component.  The first ("TFZ=") is the Z-score from the actual translation search, which depends on the accuracy of the orientation used for that search.  The second ("TFZ==") is the TFZ-equivalent, which indicates what the TFZ score would have been with the correct (refined) orientation. You should see the TFZ-equivalent is high at least for the final components of the solution, and that the LLG (log-likelihood gain) increases as each component of the solution is added.  For example, in the case of beta-blip the annotation for the single solution output in the .sol file shows these features
</p>
<pre>SOLU SET RFZ=10.7 TFZ=24.3 PAK=0 LLG=472 TFZ==24.7 RFZ=6.4 TFZ=24.4 PAK=0 LLG=1006 TFZ==29.7 LLG=1006 TFZ==29.7
SOLU 6DIM ENSE beta EULER 200.849 41.269 183.909 FRAC -0.49604 -0.15830 -0.28092 BFAC 0.00000
SOLU 6DIM ENSE blip EULER 43.749 80.793 117.292 FRAC -0.12289 0.29435 -0.09266 BFAC 0.00000
</pre>
<a name="What_to_do_in_Difficult_Cases" id="What_to_do_in_Difficult_Cases"></a><h2> <span class="mw-headline">What to do in Difficult Cases</span></h2>
<p>Not every structure can be solved by molecular replacement, but the right strategy can push the limits. What to do when the default jobs fail depends on why your structure is difficult.
</p>
<ul><li><b>Flexible Structure</b>
<dl><dd>The relative orientations of the domains may be different in your crystal than in the model. If that may be the case, break the model into separate PDB files containing rigid-body units, enter these as separate ensembles, and search for them separately. If you find a convincing solution for one domain, but fail to find a solution for the next domain, you can take advantage of the knowledge that its orientation is likely to be similar to that of the first domain. The ROTAte&nbsp;AROUnd option of the brute rotation search can be used to restrict the search to orientations within, say, 30 degrees of that of the known domain. Allow for close approach of the domains by increasing the allowed clashes with the PACK keyword by, say, 1 for each domain break that you introduce.  Alternatively, you could try generating a series of models perturbed by normal modes, with the NMAPdb keyword. One of these may duplicate the hinge motion and provide a good single model.
</dd></dl>
</li><li><b>Poor or Incomplete Model</b>
<dl><dd>Signal-to-noise is reduced by coordinate errors or incompleteness of the model. Since the rotation search has lower signal to begin with than the translation search, it is usually more severely affected. For this reason, it can be very useful to use the subsequent translation search as a way to choose among many (say 1000) orientations. THe MR_AUTO FAST search mode automatically reduces the cutoff for accepting peaks from the fast rotation function if the decault pass does not find a solution with a high z-score, but you can manually reduce this further with the PEAKS and PURGE keywords. You can also try turning off the clustering of fast rotation function peaks because the correct orientation may sit on the shoulder of a peak in the rotation function.  
</dd><dd>As shown convincingly by Schwarzenbacher <i>et al.</i> (Schwarzenbacher, Godzik, Grzechnik &amp; Jaroszewski, <i>Acta Cryst.</i> D<b>60</b>, 1229-1236, 2004), judicious editing can make a significant difference in the quality of a distant model. In a number of tests with their data on models below 30% sequence identity, we have found that Phaser works best with a "mixed model" (non-identical sidechains longer than Ser replaced by Ser).  In agreement with their results, the best models are generally derived using more sophisticated alignment protocols, such as their FFAS protocol. Use <a href="http://www.phenix-online.org/documentation/sculptor.htm" class="external text" title="http://www.phenix-online.org/documentation/sculptor.htm" rel="nofollow">phenix.sculptor</a> to edit your model.
</dd></dl>
</li><li><b>High Degree of Non-crystallographic Symmetry</b>
<dl><dd>If there are clear peaks in the self-rotation function, you can expect orientations to be related by this known NCS. Methods to automatically use such information will be implemented in a future version of Phaser. In the meantime, you can work out for yourself the orientations that would be consistent with NCS and use the  ROTAte&nbsp;AROUnd option to sample similar orientations. Alternatively, you may have an oligomeric model and expect similar NCS in the crystal. First search with the oligomeric model; if this fails, search with a monomer. If that succeeds, you can again use the ROTAte&nbsp;AROUnd option to force a subsequent monomer to adopt an orientation similar to the one you expect.
</dd></dl>
</li><li><b>What <u>not</u> to do</b>
<dl><dd>The automated mode of Phaser is fast when Phaser finds a high Z-score solution to your problem. When Phaser cannot find a solution with a significant Z-score, it "thrashes", meaning it maintains a list of 100-1000's of low Z-score potential solutions and tries to improve them. This can lead to exceptionally long Phaser runs (over a week of CPU time). Such runs are possible because the highly automated script allows many consecutive MR jobs to be run without you having to manually set 100-1000's of jobs running and keep track of the results.  "Thrashing" generally does not produce a solution: solutions generally appear relatively quickly or not at all. It is more useful to go back and analyse your models and your data to see where improvements can be made. Your system manager will appreciate you terminating these jobs.
</dd><dd>It is also not a good idea to effectively remove the packing test.  Unless there is specific evidence in the logfile that a high TF-function Z-score solution is being rejected with a few clashes, it is much better to edit the model to remove the loops than to increase the number of allowed clashes. Packing criteria are a very powerful constraint on the translation function, and increasing the number of allowed clashes beyond a few (e.g. 1-5) will increase the search time enormously without the possibility of generating any correct solutions that would not have otherwise been found.
</dd></dl>
</li><li><b>Other suggestions</b>
<dl><dd>Phaser has powerful input, output and scripting facilities that allow a large number of possibilities for altering default behaviour and forcing Phaser to do what you think it should. However, you will need to read the information in the manual below to take advantage of these facilities!
</dd></dl>
</li></ul>
<a name="How_to_Define_Data" id="How_to_Define_Data"></a><h2> <span class="mw-headline">How to Define Data</span></h2>
<p>You need to tell Phaser the name of the mtz file containing your data and the columns in the mtz file to be used using the HKLIn and LABIn keywords. Additional keywords (BINS CELL OUTLier RESOlution SPACegroup) define how the data are used.
</p>
<a name="How_to_Define_Models" id="How_to_Define_Models"></a><h2> <span class="mw-headline">How to Define Models</span></h2>
<p>Phaser must be given the models that it will use for molecular replacement. A model in Phaser is referred to as an "ensemble", even when it is described by a single file. This is because it is possible to provide a set of aligned structures as an ensemble, from which a statistically-weighted averaged model is calculated. A molecular replacement model is provided either as one or more aligned pdb files, or as an electron density map, entered as structure factors in an mtz file. Each ensemble is treated as a separate type of rigid body to be placed in the molecular replacement solution. An ensemble should only be defined once, even if there are several copies of the molecule in the asymmetric unit.
</p><p>Fundamental to the way in which Phaser uses MR models (either from coordinates or maps) is to estimate how the accuracy of the model falls off as a function of resolution, represented by the Sigma(A) curve.  To generate the Sigma(A) curve, Phaser needs to know the RMS coordinate error expected for the model and the fraction of the scattering power in the asymmetric unit that this model contributes.
</p><p>A Babinet-style correction is used to account for the effects of disordered solvent on the completeness of the model at low resolution.
</p><p>Molecular replacement models are defined with the ENSEmble keyword and the COMPosition keyword.  The ENSEmble keyword gives (amongst other things) the  RMS deviation for the Sigma(A) curve.  The COMPosition keyword is used to deduce the fraction of the scattering power in the asymmetric unit that each ensemble contributes. The composition of the asymmetric unit is defined either by entering the molecular weights or sequences of the components in the asymmetric unit, and giving the number of copies of each. Expert users can also enter the fraction of the scattering of each component directly, although the composition must still be entered for the absolute scale calculation. Please note that the composition supplied to Phaser has to include everything in the asymmetric unit, not just what is being looked for in the current search!
</p>
<a name="Building_an_Ensemble_from_Coordinates" id="Building_an_Ensemble_from_Coordinates"></a><h3> <span class="mw-headline">Building an Ensemble from Coordinates</span></h3>
<p>The RMS deviation is determined directly from RMS or indirectly from IDENtity in the ENSEmble
keyword using a formula that depends on the sequence identity and the number of residues
</p>
<table class="wikitable" style="margin-left: 30px" align="right">
<tr>
<td colspan="11" style="text-align: center;"> Initial estimate of RMS deviation
</td></tr>
<tr>
<td colspan="11" style="text-align: center;"> Number of residues versus sequence identity
</td></tr>
<tr>
<th> </th><th>    #50   </th><th>   #100  </th><th>   #200   </th><th>  #300    </th><th> #400  </th><th>   #600   </th><th>  #850 </th><th>    #1000 </th><th>   #1500 </th><th>  #2000
</th></tr>
<tr>
<td><b>ID=0%</b> </td><td>  1.579 </td><td> 1.689 </td><td> 1.875 </td><td> 2.030 </td><td> 2.164 </td><td> 2.391 </td><td> 2.625 </td><td> 2.748 </td><td> 3.093 </td><td> 3.375
</td></tr>
<tr>
<td><b>ID=10%</b> </td><td> 1.356 </td><td> 1.451 </td><td> 1.610 </td><td> 1.743 </td><td> 1.858 </td><td> 2.053 </td><td> 2.255 </td><td> 2.360 </td><td> 2.657 </td><td> 2.899
</td></tr>
<tr>
<td><b>ID=20%</b> </td><td> 1.165 </td><td> 1.246 </td><td> 1.383 </td><td> 1.497 </td><td> 1.596 </td><td> 1.764 </td><td> 1.936 </td><td> 2.027 </td><td> 2.281 </td><td> 2.489
</td></tr>
<tr>
<td><b>ID=30%</b> </td><td> 1.000 </td><td> 1.070 </td><td> 1.188 </td><td> 1.286 </td><td> 1.371 </td><td> 1.515 </td><td> 1.663 </td><td> 1.741 </td><td> 1.959 </td><td> 2.138
</td></tr>
<tr>
<td><b>ID=40%</b> </td><td> 0.859 </td><td> 0.919 </td><td> 1.020 </td><td> 1.104 </td><td> 1.177 </td><td> 1.301 </td><td> 1.428 </td><td> 1.495 </td><td> 1.683 </td><td> 1.836
</td></tr>
<tr>
<td><b>ID=50%</b> </td><td> 0.738 </td><td> 0.789 </td><td> 0.876 </td><td> 0.948 </td><td> 1.011 </td><td> 1.117 </td><td> 1.227 </td><td> 1.284 </td><td> 1.445 </td><td> 1.577
</td></tr>
<tr>
<td><b>ID=60%</b> </td><td> 0.634 </td><td> 0.678 </td><td> 0.752 </td><td> 0.814 </td><td> 0.868 </td><td> 0.959 </td><td> 1.053 </td><td> 1.103 </td><td> 1.241 </td><td> 1.354
</td></tr>
<tr>
<td><b>ID=70%</b> </td><td> 0.544 </td><td> 0.582 </td><td> 0.646 </td><td> 0.699 </td><td> 0.746 </td><td> 0.824 </td><td> 0.905 </td><td> 0.947 </td><td> 1.066 </td><td> 1.163
</td></tr>
<tr>
<td><b>ID=80%</b> </td><td> 0.467 </td><td> 0.500 </td><td> 0.555 </td><td> 0.601 </td><td> 0.640 </td><td> 0.708 </td><td> 0.777 </td><td> 0.813 </td><td> 0.915 </td><td> 0.999
</td></tr>
<tr>
<td><b>ID=90%</b> </td><td> 0.401 </td><td> 0.429 </td><td> 0.477 </td><td> 0.516 </td><td> 0.550 </td><td> 0.608 </td><td> 0.667 </td><td> 0.698 </td><td> 0.786 </td><td> 0.858
</td></tr>
<tr>
<td><b>ID=100%</b> </td><td> 0.345 </td><td> 0.369 </td><td> 0.409 </td><td> 0.443 </td><td> 0.472 </td><td> 0.522 </td><td> 0.573 </td><td> 0.600 </td><td> 0.675 </td><td> 0.737
</td></tr>
</table>
<p>The RMS deviation estimated from ID may be an underestimate of the true value if there is a slight conformational change between the model and target structures. To find a solution in these cases it may be necessary to increase the RMS from the default value generated from the ID, by say 0.5 Ångstroms. On the other hand, when Phaser succeeds in solving a structure from a model with sequence identity much below 30%, it is often found that the fold is preserved better than the average for that level of sequence identity. So it may be worth submitting a run in which the RMS error is set at, say, 1.5, even if the sequence identity is low. The table below can be used as a guide as to the default RMS value corresponding to ID.
</p><p>If you construct a model by homology modelling, remember that the RMS error you expect is essentially the error you expect from the template structure (if not worse!).  So specify the sequence identity of the template, not of the homology model.
</p><p>Only the model with the highest sequence identity is reported in the output pdb file. Also, HETATM cards in the input pdb file are ignored in the calculation of the structure factors for the ensemble, but are carried through to the output pdb file. Thus, the phases on the output mtz file (which come from the structure factors of the ensemble) do not correspond to those that would be calculated from the output pdb file, when there is more than one pdb file in an ensemble and/or the pdbfile(s) have HETATM records.
</p>
<a name="Coordinate_Editing" id="Coordinate_Editing"></a><h4> <span class="mw-headline">Coordinate Editing</span></h4>
<a name="HETATM.2FLIGANDS" id="HETATM.2FLIGANDS"></a><h5> <span class="mw-headline">HETATM/LIGANDS</span></h5>
<p>Phaser ignores the scattering from HETATM records. The HETATM records are carried though to output with occupancy set to zero. Ligands will therefore not contribute to the scattering used for molecular replacement. The exceptions to this rule are the HETATM records for MSE (seleno-methionine) MSO (seleno-methionine selenoxide) CSE (seleno-cysteine) CSO (seleno-cysteine selenoxide) ALY (acetyllysine) MLY (n-dimethyl-lysine) and MLZ (n-methyl-lysine) which are used in the scattering and carried through to output with their original occupancy. If you wish to include any HETATM records in the scattering the record name use the keyword ENSE modlid HETATOM ON
</p>
<a name="WATER" id="WATER"></a><h5> <span class="mw-headline">WATER</span></h5>
<p>Water molecules (identified by the residue name OW WAT HOH H2O OH2 MOH WTR or TIP) are deleted from the pdb file on input, are not used in the scattering and are not carried through to file output. If you want to retain water molecules you will need to change the residue name to something other than this (e.g. WWW) so that the atoms are not identified as water. To include the water molecules in the scattering, the HETATM records will also have to be changed to ATOM records as described above.
</p>
<a name="Building_an_Ensemble_from_Electron_Density" id="Building_an_Ensemble_from_Electron_Density"></a><h3> <span class="mw-headline">Building an Ensemble from Electron Density</span></h3>
<p>When using density as a model, it is necessary to specify both the extent (x,y,z limits) of the cut-out region of density, and the centre of this region. With coordinates, Phaser can work this out by itself. This information is needed, for instance, to decide how large rotational steps can be in the rotation search and to carry out the molecular transform interpolation correctly. In the case of electron density, the RMS value does not have the same physical meaning that it has when the model is specified by atomic coordinates, but it is used to judge how the accuracy of the calculated structure factors drops off with resolution. A suitable value for RMS can be obtained, in the case of density from an experimentally-phased map, by choosing a value that makes the SigmaA curve fall off with resolution similarly to the mean figures-of-merit. In the case of density from an EM image reconstruction, the RMS value should make the SigmaA curve fall off similarly to a Fourier correlation curve used to judge the resolution of the EM image.
</p><p>For detailed information, including a tutorial with example scripts, see
<a href="../../../../articles/u/s/i/Using_Electron_Density_as_a_Model_13c4.html" title="Using Electron Density as a Model"> Using density as a model</a>
</p>
<a name="How_to_Define_Composition" id="How_to_Define_Composition"></a><h2> <span class="mw-headline">How to Define Composition</span></h2>
<p>The composition defines the total amount of protein and nucleic acid that you have in the asymmetric unit not the fraction of the asymmetric unit that you are searching for.
</p>
<a name="Default_Composition" id="Default_Composition"></a><h3> <span class="mw-headline">Default Composition</span></h3>
<p>For convenience, the composition defaults to 50% protein scattering by volume (the average for protein crystals). It is better to enter it explicitly, even if only to check that you have correctly deduced the probable content of your crystal. If your crystal has higher or lower solvent content than this, or contains nucleic acid, then the composition should be entered explicitly.
</p>
<a name="Composition_by_Solvent_Content" id="Composition_by_Solvent_Content"></a><h3> <span class="mw-headline">Composition by Solvent Content</span></h3>
<p>Scattering is determined from the solvent content of the crystal, assuming that the crystal contains protein only, and the average distribution of amino acids in protein. If your crystal contains nucleic acid or your protein has an unusual amino acid distribution then the composition should be entered explicitly using the MW or sequence options.
</p>
<a name="Composition_by_Number_of_Residues_in_ASU" id="Composition_by_Number_of_Residues_in_ASU"></a><h3> <span class="mw-headline">Composition by Number of Residues in ASU</span></h3>
<p>Scattering is determined from the number of residues in the asymmetric unit, assuming that the crystal contains protein only or nucleic acid only, and assuming an average distribution of residues for either. If your crystal contains a mixture then the composition should be entered explicitly using the MW or sequence options.  If your crystal has an unusual residue distribution then the composition should be entered explicitly using the sequence options.
</p>
<a name="Composition_by_Molecular_Weight" id="Composition_by_Molecular_Weight"></a><h3> <span class="mw-headline">Composition by Molecular Weight</span></h3>
<p>The composition is calculated from the molecular weight of the protein and nucleic acid assuming the protein and nucleic acid have the average distribution of amino acids and bases. If your protein or nucleic acid has an unusual amino acid or base distribution the composition should be entered by sequence. You can mix compositions entered by molecular weight with those entered by sequence.
</p>
<a name="Composition_by_Sequence" id="Composition_by_Sequence"></a><h3> <span class="mw-headline">Composition by Sequence</span></h3>
<p>The composition is calculated from the amino acid sequence of the protein and the base sequence of the nucleic acid in fasta format. You can mix compositions entered by molecular weight with those entered by sequence. Individual atoms can be added to the composition with the COMPOSITION ATOM keyword. This allows the explicit addition of heavy atoms in the structure e.g. Fe atoms.
</p>
<a name="Composition_by_Percentage_Scattering" id="Composition_by_Percentage_Scattering"></a><h3> <span class="mw-headline">Composition by Percentage Scattering</span></h3>
<p>The fraction scattering of each ensemble can be entered directly.  The fraction scattering of each ensemble is normally automatically worked out from the average scattering from each ensemble (calculated from the pdb files if entered as coordinates, or from the protein and nucleic acid molecular weights if entered as a map) divided by the total scattering given by the composition, but entering the fraction scattering directly overrides this calculation. This option is for use when the pdb files of the models in the ensemble are unusual e.g.  consist only of C-alpha atoms, or only of hydrogen atoms (as in the CLOUDS method for NMR).
</p>
<a name="How_to_Define_Solutions" id="How_to_Define_Solutions"></a><h2> <span class="mw-headline">How to Define Solutions</span></h2>
<p>Phaser writes out files ending in ".sol" and ".rlist" that contain the solution information from the job. The root of the files is given by the ROOT keyword. By default, the root filename is PHASER. These files can be read back into subsequent runs of Phaser to build up solutions containing more than one molecule in the asymmetric unit.
</p><p>"PHASER.sol"  files are generated by all modes (rotation function modes with VERBOSE output), and contain the current idea of potential molecular replacement solutions.
</p><p>"PHASER.rlist"  files are generated by the rotation function modes, and are used as input for performing translation functions. (They are also produced by degenerate (2D) translation functions, for performing a translation function to find the third dimension)
</p><p>For simple MR cases you don't really need to know how to define molecular replacement solutions.  However, for difficult cases you might need to edit the files "PHASER.sol" and "PHASER.rlist" files manually
</p>
<a name=".22sol.22_Files" id=".22sol.22_Files"></a><h3> <span class="mw-headline"> "sol" Files</span></h3>
<p>At different stages of molecular replacement, an Ensemble will be oriented but not positioned (after the rotation search), or oriented and positioned (after the translation search), or, rarely, oriented and the position in 2 of 3 dimensions known. These three states correspond to solutions defined by the keywords SOLUtion 3DIM, SOLUtion 6DIM, and SOLUtion 5DIM. Each Ensemble in the asymmetric unit has its own SOLUtion keyword. Solutions of the type 3DIM are given by the rotation function, solutions of the type 6DIM are given by the translation function, and solutions of the type 5DIM are given by the degenerate translation function.
</p><p>When more than one (potential) molecular replacement solution is present, the solutions are separated with the SOLUTION SET keywords.
</p>
<a name=".22rlist.22_Files" id=".22rlist.22_Files"></a><h3> <span class="mw-headline">"rlist" Files</span></h3>
<p>These files define a rotation function list. The peak list is given with a series of SOLUtion TRIAl keywords.
</p><p>If a partial solution is already known, then the information for the currently "known" parts of the asymmetric unit is given in the form used for the PHASER.sol file, followed by the list of trial orientations for which a translation function is to be performed.
</p><p>If a degenerate translation function is performed, then a SOLUtion TRIAl line is produced with the degenerate translation information present, ready for performing the translation function on the third dimension.
</p>
<a name="Fixed_partial_structure" id="Fixed_partial_structure"></a><h3> <span class="mw-headline">Fixed partial structure</span></h3>
<p>If you have the coordinates of a partial solution with the pdb coordinates of the known structure in the correct orientation and position, then you can force Phaser to use these coordinates. Use the SOLUTION keyword to fix a rotation of 0 0 0 and a position of 0 0 0 for these coordinates.
</p>
<a name="How_to_Select_Peaks" id="How_to_Select_Peaks"></a><h2> <span class="mw-headline">How to Select Peaks</span></h2>
<p>The selection of peaks saved for output in the rotation and translation functions can be done in four different ways.
</p>
<ul><li><b>Select by Percentage</b>
<dl><dd> Percentage of the top peak, where the value of the top peak is defined as 100% and the value of the mean is defined as 0%.
</dd><dd> Default, cutoff=75%. This criteria has the advantange that at least one peak (the top peak) always survives the selection. If the top solution is clear, then only the one solution will be output, but if the distribution of peaks is rather flat, then many peaks will be output for testing in the next part of the MR procedure (e.g. many peaks selected from the rotation function for testing with a translation function).   
</dd></dl>
</li><li><b>Select by Z-score</b>
<dl><dd> Number of standard deviations (sigmas) over the mean (the Z-score). 
</dd><dd> Absolute significance test. Not all searches will produce output if the cutoff value is too high (e.g. 5 sigma).   
</dd></dl>
</li><li><b>Select by Number</b>
<dl><dd> Number of top peaks to select. 
</dd><dd> If the distribution is very flat then it might be better to select a fixed large number (e.g. 1000) of top rotation peaks for testing in the translation function.
</dd></dl>
</li><li><b>No selection</b>
<dl><dd> All peaks are selected. 
</dd><dd> Enables full 6 dimensional searches, where all the solutions from the rotation function are output for testing in the translation function. This should never be necessary; it would be much faster and probably just as likely to work if the top 1000 peaks were used in this way.
</dd></dl>
</li></ul>
<div class="thumb tnone"><div class="thumbinner" style="width:102px;"><a href="../../../../articles/p/h/a/File%7EPhaser_selection.gif_06b9.html" class="image" title="Selection criteria"><img alt="" src="../../../../images/thumb/3/3b/Phaser_selection.gif/100px-Phaser_selection.gif" width="100" height="50" border="0" class="thumbimage" /></a>  <div class="thumbcaption"><div class="magnify"><a href="../../../../articles/p/h/a/File%7EPhaser_selection.gif_06b9.html" class="internal" title="Enlarge"><img src="../../../../skins/common/images/magnify-clip.png" width="15" height="11" alt="" /></a></div>Selection criteria</div></div></div>
<p>Peaks can also be clustered or not clustered prior to selection in steps 1 and 2.
</p>
<ul><li><b>Clustering Off</b>
</li></ul>
<dl><dd> All high peaks on the search grid are selected
</dd></dl>
<ul><li><b>Clustering On</b>
</li></ul>
<dl><dd> Points on the search grid with higher neighbouring points are removed from the selection
</dd></dl>
<p><br />
</p>
<div class="thumb tnone"><div class="thumbinner" style="width:102px;"><a href="../../../../articles/p/h/a/File%7EPhaser_clustering.gif_ef53.html" class="image" title="Clustering"><img alt="" src="../../../../images/thumb/e/ef/Phaser_clustering.gif/100px-Phaser_clustering.gif" width="100" height="54" border="0" class="thumbimage" /></a>  <div class="thumbcaption"><div class="magnify"><a href="../../../../articles/p/h/a/File%7EPhaser_clustering.gif_ef53.html" class="internal" title="Enlarge"><img src="../../../../skins/common/images/magnify-clip.png" width="15" height="11" alt="" /></a></div>Clustering</div></div></div>
<a name="How_to_Control_Output" id="How_to_Control_Output"></a><h2> <span class="mw-headline">How to Control Output</span></h2>
<p>The output of Phaser can be controlled with optional keywords. 
</p><p>The ROOT keyword is not compulsory (the default root filename is "PHASER"), but should always be given, so that your jobs have separate and meaningful output filenames.
</p><p>The TOPFiles keyword controls the number of potential MR solutions for which PDB and (in the appropriate modes) MTZ files are produced.
</p><p>For the MR_AUTO, MR_RNP and MR_LLG modes, unless HKLOut OFF is given as an optional keyword, Phaser produces an MTZ file with "SigmaA" type weighted Fourier map coefficients for producing electron density maps for rebuilding.
</p>
<table class="wikitable" style="text-align:left" width="100%">

<tr>
<th> MTZ Column Labels </th><th> Description
</th></tr>
<tr>
<td> FWT/PHWT </td><td> Amplitude and phase for 2<i>m</i>&#124;F<sub>obs</sub>&#124;-<i>D</i>&#124;F<sub>calc</sub>&#124; exp(<i>i</i>&alpha;<sub>calc</sub>) map
</td></tr>
<tr>
<td> DELFWT/PHDELWT </td><td> Amplitude and phase for <i>m</i>&#124;F<sub>obs</sub>&#124;-<i>D</i>&#124;F<sub>calc</sub>&#124; exp(<i>i</i>&alpha;<sub>calc</sub>) map
</td></tr>
<tr>
<td> FOM </td><td> <i>m</i>, analogous to the "Sim" weight, to estimate the reliability of &alpha;<sub>calc</sub>
</td></tr>
<tr>
<td> HLA/HLB/HLC/HLD </td><td> Hendrickson-Lattman coefficients encoding the phase probability distribution
</td></tr></table>

<!-- 
NewPP limit report
Preprocessor node count: 28/1000000
Post-expand include size: 0/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->
<div class="printfooter">
</div>
	    	    <!-- end content -->
	    <div class="visualClear"></div>
	  </div>
	</div>
      </div>
      <div id="column-one">
	<div class="portlet" id="p-logo">
            <a style="background-image: url(../../../../misc/PhaserwikiLogo.png);"
	    href="http://www.phaser.cimr.cam.ac.uk/index.php/Phaser_Crystallographic_Software"
	    title="Phaser Crystallographic Software"></a>
	</div>
	<script type="text/javascript"> if (window.isMSIE55) fixalpha(); </script>
		<div class='portlet' id='p-site'>
	  <h5>site</h5>
	  <div class='pBody'>
	    <ul>
	    	      <li id="n-Home"><a href="../../../../index.html">Home</a></li>
	     	      <li id="n-News"><a href="../../../../articles/n/e/w/News.html">News</a></li>
	     	      <li id="n-Events"><a href="../../../../articles/e/v/e/Events.html">Events</a></li>
	     	      <li id="n-Downloads"><a href="../../../../articles/d/o/w/Downloads.html">Downloads</a></li>
	     	    </ul>
	  </div>
	</div>
		<div class='portlet' id='p-documentation'>
	  <h5>documentation</h5>
	  <div class='pBody'>
	    <ul>
	    	      <li id="n-Manuals"><a href="../../../../articles/m/a/n/Manuals.html">Manuals</a></li>
	     	      <li id="n-MR-Phasing"><a href="../../../../articles/m/o/l/Molecular_Replacement_a23a.html">MR Phasing</a></li>
	     	      <li id="n-SAD-Phasing"><a href="../../../../articles/e/x/p/Experimental_Phasing_5400.html">SAD Phasing</a></li>
	     	      <li id="n-Keyword-Examples"><a href="../../../../articles/m/r/_/MR_using_keyword_input_9c88.html">Keyword Examples</a></li>
	     	      <li id="n-Python-Examples"><a href="../../../../articles/p/y/t/Python_Example_Scripts_7723.html">Python Examples</a></li>
	     	      <li id="n-Tutorials"><a href="../../../../articles/t/u/t/Tutorials.html">Tutorials</a></li>
	     	      <li id="n-FAQ"><a href="../../../../articles/f/a/q/FAQ_1fe9.html">FAQ</a></li>
	     	      <li id="n-Top-Ten-Tips"><a href="../../../../articles/t/o/p/Top_Ten_Tips_3fcd.html">Top Ten Tips</a></li>
	     	      <li id="n-Bugs"><a href="../../../../articles/b/u/g/Bugs.html">Bugs</a></li>
	     	      <li id="n-Publications"><a href="../../../../articles/p/u/b/Publications.html">Publications</a></li>
	     	      <li id="n-View-SVN"><a href="../../../../articles/s/v/n/SVN_Repository_93a5.html">View SVN</a></li>
	     	    </ul>
	  </div>
	</div>
		<div class='portlet' id='p-other'>
	  <h5>other</h5>
	  <div class='pBody'>
	    <ul>
	    	      <li id="n-Contact"><a href="../../../../articles/c/o/n/Contact.html">Contact</a></li>
	     	      <li id="n-Developers"><a href="../../../../articles/d/e/v/Developers.html">Developers</a></li>
	     	      <li id="n-Licences"><a href="../../../../articles/l/i/c/Licences.html">Licences</a></li>
	     	      <li id="n-External-Links"><a href="../../../../articles/e/x/t/External_Links_ff0e.html">External Links</a></li>
	     	      <li id="n-Help"><a href="../../../../articles/c/o/n/Help%7EContents_22de.html">Help</a></li>
	     	    </ul>
	  </div>
	</div>
	<!--
	<div id="p-search" class="portlet">
	  <h5><label for="searchInput">Search</label></h5>
	  <div id="searchBody" class="pBody">
	    <form action="javascript:goToStatic(3)" id="searchform"><div>
	      <input id="searchInput" name="search" type="text"
	        accesskey="f" value="" />
	      <input type='submit' name="go" class="searchButton" id="searchGoButton"
	        value="Go" />
	    </div></form>
	  </div>
	</div>
-->
	      </div><!-- end of the left (by default at least) column -->
      <div class="visualClear"></div>
      <div id="footer">
<!--
    <div id="f-poweredbyico"><a href="http://www.mediawiki.org/"><img src="../../../../skins/common/images/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" /></a></div>	-->
<!-- date DATE Date -->
	<ul id="f-list">
	   Static <a href="http://www.phaser.cimr.cam.ac.uk/index.php/Phaser_Crystallographic_Software">Phaserwiki</a> 
Thu Jul 25 23:20:16 BST 2013
<!--
	            </li>
	  	  <li id="f-credits">This page was last modified 14:04, 17 July 2013 by <a href="../../../../articles/r/d/o/User%7ERdo20_c026.html" class="new" title="User:Rdo20 (page does not exist)">Rob Oeffner</a>. Based on work by <a href="../../../../articles/a/i/r/User%7EAirlie_8634.html" title="User:Airlie">Airlie McCoy</a>,&#32;<a href="../../../../articles/r/a/n/User%7ERandy_8422.html" title="User:Randy">Randy Read</a>&#32;and&#32;<a href="../../../../articles/w/i/k/User%7EWikiSysop_5eaf.html" title="User:WikiSysop">WikiSysop</a>.</li>	  	  <li id="f-about"><a href="../../../../articles/a/b/o/Phaserwiki%7EAbout_932f.html" title="Phaserwiki:About">About Phaserwiki</a></li>	  	  -->
	</ul>
      </div>
    </div>
  </body>
</html>
