<html><head>
   <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
   <meta name="Author" content="Martyn"><title>CCP4 Tutorial - Molecular Replacement</title></head>
<body>

<h1 align="center">CCP4 Tutorial: Molecular Replacement</h1>

<p>We will solve the hypF structure by molecular replacement, using 
several programs and approaches. Other MR examples can be found at the end of this tutorial, and at:
<dl>
<dt></dt><dd><a href="http://www.ccp4.ac.uk/dist/examples/tutorial/html/mr-tutorial.html">CCP4
tutorial</a> (available in CCP4 suite). This uses a cardiotoxin 1tgx.
</dd><dt></dt><dd><a href="http://www.ysbl.york.ac.uk/%7Ealexei/downloads/tutorial_MR.tar.gz">Molrep 
tutorial</a> (Alexei Vagin).
</dd><dt></dt><dd><a href="http://www.phaser.cimr.cam.ac.uk/index.php/Tutorials">
Phaser tutorial</a> (Cambridge).
</dd></dl>
</p>
<p>
When this tutorial is obtained as part of the CCP4 distribution, $MR_TUTORIAL corresponds
to $CCP4/examples/mr_tutorial_2006 <br>
<b>Note</b> that it has been updated since 2006!
</p>


<h2>Checking the data</h2>

<p>Target is the acylphosphatase-like domain of hydrogenase maturation factor HypF
from E.coli, see Rosano et al, <em>JMB</em>, <b>321</b>, 785 (2002). HypF-ACP sulphate 
and phosphate complexes deposited as 1gxt and 1gxu respectively. </p>

<p>This protein has a Hg derivative. You may have processed this data in a preceding tutorial.
We have prepared a reflection file for you including the data from 1gxu, 1gxt, the Hg derivative, 
and some experimental phases based on the Hg sites. </p>

<p>There is native data in H32 to 1.3 A resolution. The target has 91 residues and a Matthews 
calculation strongly suggests only one molecule in the asymmetric unit.</p>

<p>We first use <b>Sfcheck</b> to check a few things about the data:
</p><ol>
<li>Select the <font color="#ff0000">Program List</font> module and open the 
<font color="#ff0000">sfcheck</font> task window.
</li><li>Enter a title.
</li><li>Un-check <font color="#009900">Run Rampage to analyse structure geometry</font>
and <font color="#009900">Run Procheck to analyse structure geometry</font>
(we do not yet have any coordinates)
</li><li>Select <font color="#009900">Run Sfcheck to analyse</font>
<font color="#ff0000">experimental data only</font>
</li><li>Enter <font color="#009900">MTZ in</font> 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/hypF-1gxu-1gxt-HG_scaleit1.mtz</font> 
and select the labels <font color="#009900">F</font> <font color="#ff0000">FP1gxu</font>, 
<font color="#009900">SIGF</font> <font color="#ff0000">SIGFP1gxu</font> and
<font color="#009900">Free</font> <font color="#ff0000">FREE</font>
</li><li>Check that a suitable filename has been generated for 
<font color="#009900">Sfcheck Output PS</font> 
</li><li>Keep all defaults, and click <font color="#ff0000">Run -&gt; Run Now</font>.
</li></ol>
<p></p>

<p>Sfcheck produces a postscript file with some useful things (see under 
<font color="#ff0000">View Files from Job</font>):
</p><ul>
<li>Anisotropy of data (it is not very anisotropic)
</li><li>Overall B from Wilson plot of 21.9 A**2
</li><li>Pseudo-translation not detected (from analysis of the native Patteron map)
</li></ul>
<p></p>

<p>Also check the log file <font color="#ff0000">View Files from Job</font> then
<font color="#ff0000">View Log File</font>:
</p><ul>
<li>This includes the results of the a twinning test:
<pre>  Perfect twinning test &lt;I^2&gt; / &lt;I&gt;^2 :  2.0573
</pre>
A value of 2.0 indicates untwinned data, whereas perfectly twinned data would
have a second moment of 1.5
</li></ul>
<p></p>


<h2>Choice of search models</h2>
<p>The target is an acylphosphatase-like domain. A search of the PDB reveals two
acylphosphatases with a sequence identity to the target of about 31% - 1v3z and 1w2i.
Each has two chains in the asymmetric unit, either of which could be used as
the basis of a search model.</p>

<p>Normally you would use something like Chainsaw at this point to prepare a search 
model from the template. As an exercise, we are going to try MR straightaway. We will
return to Chainsaw later before running Phaser.</p>

<h4>Notes on Sequence Alignment </h4>

<p>There are many ways of approaching this, and the different tools 
will give slightly different assessments.
The sequence identity depends on the definitions used (i.e. treatment of gaps and alignment
length), the specific alignment technique, and whether bits have been chopped
out of the model. </p>

<h2>Molrep</h2>

<h3>Run 1</h3>

<p>We will use chain B of 1v3z as the search model (file $MR_TUTORIAL/data/hypF/1v3z_B.pdb). 
</p><ol>
<li>Select the <font color="#ff0000">Molecular Replacement</font> module and open the 
<font color="#ff0000">Run Molrep - auto MR</font> task window.
</li><li>Enter a title.
</li><li><font color="#009900">Do</font> <font color="#ff0000">molecular replacement</font> 
<font color="#009900">performing</font> <font color="#ff0000">rotation and translation function</font> 
should be already selected.
</li><li>Enter the MTZ file <font color="#ff0000">$MR_TUTORIAL/data/hypF/hypF-1gxu-1gxt-HG_scaleit1.mtz</font> 
and select the labels <font color="#009900">FP</font> <font color="#ff0000">FP1gxu</font> and 
<font color="#009900">SIGFP</font> <font color="#ff0000">SIGFP1gxu</font>
</li><li>Enter the PDB file <font color="#ff0000">$MR_TUTORIAL/data/hypF/1v3z_B.pdb</font>
</li><li>(Optional) You can use an upper resolution cut off of 3A to speed up the calculation,
see folder <font color="#009900">Experimental Data</font>.
</li><li>Keep all defaults, and click <font color="#ff0000">Run -&gt; Run Now</font>.
</li></ol>
When the job has finished, look at the log file (View Files from Job -&gt; View Log File).
Note the following:
<ul>
<li>Molrep automatically estimates:
<pre>   
    INFO: expected number of models :    1  
    INFO: V_model:   61.4% (of asymm. part of u.c.) 
</pre>
which is correct. The estimate may be unreliable when there are many monomers in 
the asymmetric unit, in which case it can be set explicitly with the keyword NMON
(see folder <font color="#009900">Search Parameters</font> in the Molrep GUI).
</li><li>Molrep checks whether or not an anisotropy correction is necessary:
<pre>   INFO: Anisotropicy will not be used
</pre>
</li><li>The first table is a list of peaks of the Cross Rotation Function (CRF), sorted according to their 
heights. This is followed by a plot showing which peaks are related.
</li><li>The second table shows the best Translation Function (TF) for each of the CRF peaks (scored
according to the correlation coefficient * PKmax). Other TF solutions can be viewed in the file
View Files from Job -&gt; Output Files ... &lt;proj_dir&gt;_&lt;job_no&gt;_molrep.doc
</li><li>The final table gives a list of solutions, sorted according to the score.
</li><li>Molrep reports a contrast of 3.90 (the precise value you get will depend on 
Molrep version, resolution limits used, etc.) and stops. This contrast value suggests a correct 
solution.
</li></ul>
<p></p>

<h3>Run 2</h3>

<p>
In fact, we can make use of our knowledge of the target, and this will often improve the 
solution. The search model has a moderately low sequence 
identity with the target and therefore the majority of the side chains are incorrect. Molrep 
can make use of the target sequence to improve the search model.
</p><ol>
<li>Select the previous job, and click <font color="#ff0000">ReRun Job</font>
</li><li>Most of the parameters should be set correctly, but you should change the title, and the name of
the <font color="#009900">Coords out</font> file, so that it is different from the first job.
</li><li>This time, select <font color="#009900">Use sequence</font> in the protocol section. A folder 
will open below where you can specify the name of the target sequence file 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/hypF_Ndom.seq</font>
</li><li><font color="#ff0000">Click Run -&gt; Run Now</font>
</li></ol>
<p></p>

<p>Look at the log file of this job.
</p><ul>
<li>After a section about the input MTZ file, there are details of the sequence alignment
between the target sequence you have supplied and the sequence of the search model (i.e. the
PDB file).
</li><li>Molrep reports a sequence identity of about 30%. This is lower than other estimates
because Molrep is more conservative in introducing gaps into the alignment.
</li><li>Molrep outputs tables for the CRF and TF as before.
</li><li>The contrast is now 3.98 and Molrep finishes. In this case, the improvement in the MR
solution is only marginal, although the it will probably form a better starting point for
building the target structure.
</ul>
<p></p>

<h3>Checking the solution</h3>

<p>The top MR solution is applied to the input coordinates, and the positioned PDB file is
written out as <font color="#ff0000">1v3z_B_molrep2.pdb</font>. The contrast indicates
that this is probably a correct solution, but this should be checked!</p>

<p>The positioned model can be submitted for a few cycles of automated refinement,
then checked manually against 2mFo-DFc and mFo-DFc maps, using a
graphics program such as Coot. Since we have a good resolution dataset,
the model can also be passed to ARP/wARP for rebuilding. Refinement, validation 
and model re-building are covered in other tutorials.</p>

<p>In fact, the Molrep solution is related to the deposited structure 1gxu by
the symmetry operation -Y+2/3,  X-Y+1/3,  Z+1/3.   Comparison of the structures
in CCP4mg shows that the beta sheet and one of the two helices are well matched,
but there are significant differences elsewhere. </p>

<p>In general, if we want to compare an MR solution to the deposited structure, 
then we need to take into account possible symmetry
operations and possible changes of origin. Two solutions may be identical, even if
it is not obvious from a quick look in a graphics program. This can be checked with
the csymmatch utility:
<ol>
<li>Select the <font color="#ff0000">Symmetry match models</font> task in 
module <font color="#ff0000">Coordinate Utilities</font>. 
<li>Enter the MR solution PDB file
as the <font color="#009900">Work PDB in</font>, and the deposited structure (1gxu) as
<font color="#009900">Reference PDB in</font>. 
<li>Select <font color="#009900">Apply
origin shift and hand correction</font> and run.
</ol>
The log file reports the symmetry operator and change of origin which give the best
match, and a normalised score for the match is reported. The output PDB file has this
transformation applied, and can be compared to the reference PDB file.
<em>Of course, usually we don't have a deposited structure to compare with, but the
same process is useful to compare different MR solutions.</em>
</p>

<h2>Chainsaw</h2>

<p>Search models can also be prepared using Chainsaw. Chainsaw takes an external
sequence alignment, which can be generated by many bioinformatics tools and/or manually
adjusted. In this job, we will create a model based on chain B of 1v3z, using a
previously prepared alignment to the target. 
</p><ol>
<li>Select the <font color="#ff0000">Molecular Replacement</font> module and open the 
<font color="#ff0000">Create Search Model</font> task window in the 
<font color="#ff0000">Model Generation</font> folder.
</li><li>Enter a title.
</li><li>Leave <font color="#009900">Create search model</font> 
<font color="#ff0000">using Chainsaw</font> unchanged.
</li><li>Leave <font color="#009900">Prune non-conserved residues to</font> 
<font color="#ff0000">gamma atom</font> unchanged.
</li><li>Enter <font color="#009900">PDB in</font> 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/1v3z_B.pdb</font>
</li><li>Use the sequence alignment format  <font color="#ff0000">PIR</font> and
enter the file <font color="#009900">Alignment in</font> 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/1v3z_B_to_target.pir</font>
</li><li><font color="#ff0000">Click Run -&gt; Run Now</font>
</li></ol>
<p></p>
<p>Chainsaw produces a coordinate file <font color="#ff0000">1v3z_B_chainsaw1.pdb</font>
which is an edited version of the input PDB file. 6 residues that do not align to the
target sequence have been deleted. Of the rest, 34 have been left unchanged and 50 have
had their side chains cut back to the gamma atom. The output PDB file uses the naming
and numbering of the target sequence.
</p>
<p>Have a look at the log file:
</p><ul>
<li>At the top, the alignment used is confirmed.
</li><li>Then there is a listing of all the model residues, with the action applied (deleted,
conserved, mutated). 
</li><li>Finally, there is a summary of the changes made. This includes the estimated
sequence identity. Note that this is not unique, but depends on the particular sequence
alignment used.
</li></ul>
<p> Now repeat this exercise using the other search model, based on chain A of 1w2i.
We can overlap the two models and use the ensemble as input to Phaser (in place of
individual search models).

</p><li>Enter <font color="#009900">PDB in</font> 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/1w2i_A.pdb</font>
</li><li>Use the sequence alignment format  <font color="#ff0000">PIR</font> and
enter the file <font color="#009900">Alignment in</font> 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/1w2i_A_to_target.pir</font>
 
<p></p>

<h2>Aligning the models</h2>

<p> These models can be aligned and the overlapped structures used as input to Phaser.
</p><ol>
<li>Select the <font color="#ff0000">Coordinate Utilities </font> module and open the 
<font color="#ff0000">Superpose Molecules</font> task window.
</li><li>Enter a title.
</li><li>Change mode to <font color="#009900">Superpose using Secondary Structure Matching</font>.
</li><li>Enter <font color="#009900">Moving</font> 
<font color="#ff0000">1w2i_A_chainsaw1.pdb</font>
</li><li>Enter <font color="#009900">Fixed</font> 
<font color="#ff0000">1v3z_B_chainsaw1.pdb</font>
</li><li>Enter <font color="#009900">PDB out</font> 
<font color="#ff0000">1w2i_A_to_1v3z_B_chainsaw1.pdb</font>
</li><li><font color="#ff0000">Click Run -&gt; Run Now</font>
</li></ol>
<p> The 1w2i_A_chainsaw1.pdb has been moved to overlap 1v3z_B_chainsaw1.pdb.
The log file shows the transformation used, and gives an RMSD = 0.305 A between 
84 C-alpha atoms of the superposed structures.</p>

<h2>Phaser</h2>
<p>Using the superposed search models generated by Chainsaw, we will now use Phaser to solve hypF.
Phaser is designed to use ensembles of models to improve the signal.</p><ol>
<li>Select the <font color="#ff0000">Molecular Replacement</font> module and open the 
<font color="#ff0000">Run Phaser</font> task window.
</li><li>Enter a title.
</li><li>Leave <font color="#009900">Mode for molecular replacement</font> 
<font color="#ff0000">automated search</font> unchanged.
</li><li>Enter <font color="#009900"></font> 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/hypF-1gxu-1gxt-HG_scaleit1.mtz</font>
and select the labels <font color="#009900">FP</font> <font color="#ff0000">FP1gxu</font> and 
<font color="#009900">SIGFP</font> <font color="#ff0000">SIGFP1gxu</font>
</li><li>In the folder <font color="#009900">Define ensembles ...</font>, enter
the <font color="#009900">PDB #1</font> 
<font color="#ff0000">1v3z_B_chainsaw1.pdb</font>.
Set the similarity to be <font color="#009900">sequence identity</font> 
<font color="#ff0000">0.38</font>
</li><li>To add another model click  <font color="#009900">Add superimposed PDB file to the ensemble</font>, enter
the <font color="#009900">PDB #2</font> 
<font color="#ff0000">1w2i_A_to_1v3z_B_chainsaw1.pdb</font>.
Set the similarity to be <font color="#009900">sequence identity</font> 
<font color="#ff0000">0.38</font>
</li><li>In the folder <font color="#009900">Define composition of the asymmetric unit</font>, 
select <font color="#009900">Total scattering determined by</font>
<font color="#009900">components in asymmetric unit</font>, and
enter the <font color="#009900">SEQ file</font> 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/hypF_Ndom.seq</font>
and leave <font color="#009900">Number in asymmetric unit</font> 
<font color="#ff0000">1</font> unchanged.
</li><li>In the folder <font color="#009900">Search parameters</font>, select
<font color="#009900">Perform search using</font> 
<font color="#ff0000">ensemble1</font>
</li><li><font color="#ff0000">Click Run -&gt; Run Now</font>
</li></ol>
<p></p>
<p>Have a look at the log file:
</p><ul>
<li>After details about the input parameters, there is information on the anisotropy
correction used (compare to the output of Sfcheck above). This is followed by
a Matthews coefficient calculation.
</li><li>Phaser then calculates a Fast Rotation Function (FRF). It finds 8 peaks greater
than 75% of the top peak (this threshold can be changed with the option
<font color="#009900">Rotation search peak selection</font> in the folder
<font color="#009900">Additional parameters</font>).
</li><li>These peaks are passed to the Fast Translation Function (FTF). Detailed results
for each rotation peak are given, followed by a summary table: 
<font color="#009900">Beware - these numbers may differ slightly for different versions of Phaser.</font>
<pre>  

   Fast Translation Function Table: Space Group H 3 2
   --------------------------------------------------
   #SET #TRIAL      Top    (Z)    Second    (Z)     Third    (Z)    Ensemble
      1      1    24.40 ( 4.70)    23.05 ( 4.55)    21.50 ( 4.37)   ensemble1
      1      2    44.15 ( 6.97)        -      -         -      -    ensemble1
      1      3    21.54 ( 4.75)    20.32 ( 4.61)    19.96 ( 4.56)   ensemble1
      1      4    19.88 ( 4.34)    17.71 ( 4.10)    17.71 ( 4.10)   ensemble1
      1      5    25.10 ( 5.11)    24.40 ( 5.03)    24.24 ( 5.01)   ensemble1
      1      6    19.60 ( 4.57)    19.33 ( 4.54)    18.65 ( 4.47)   ensemble1
      1      7    17.73 ( 4.75)    15.86 ( 4.54)    15.56 ( 4.50)   ensemble1
      1      8    17.68 ( 4.75)    15.08 ( 4.45)    13.31 ( 4.24)   ensemble1
    ---- ------

</pre>
The second trial (based on the 2nd peak of the FRF) gives a clear solution, with
a good Z-score, and a single significant peak of the FTF.
</li><li>Next is a check on packing for this good solution. Phaser finds 2 clashes between
a C-alpha and a C-alpha of a symmetry-related molecule. Because the
threshold is set to 10 clashes, this solution is accepted.
</li><li>Finally, Phaser refines the MR solution, and displays the improvement in the
log-likelihood gain (LLG).
</li><li>Phaser outputs a .sol file containing the MR solution, a .pdb file containing
the correctly positioned model, and .mtz file containing the original data plus
a calculated structure factor from the model and columns of map coefficients.
</li></ul>
<p></p>
<p>Checking the solution:
</p><ul>
<li>In fact, the Phaser solution is related to the deposited structure 1gxu by
the symmetry operation -X+Y+2/3,  -X+1/3,  Z+1/3   We are lucky - the spacegroup H32 has 
two possible origins (see <font color="#ff0000">$CHTML/alternate_origins.html</font>),
and the solution could equally well have been on the origin (0.0,0.0,0.5). 
Comparison of the structures
in CCP4mg shows that the beta sheet and one of the two helices are well matched,
but there are significant differences elsewhere.<br>
</li><li>The .pdb and .mtz files from Phaser can be inspected directly in Coot.
This shows good agreement in most places, but also highlights problem areas.
</li><li>Do 20 cycles of restrained refinement in REFMAC (<font color="#ff0000">Run
Refmac5</font> task in module <font color="#ff0000">Refinement</font>).
</li><li>Run ACORN to remove all phase biase (<font color="#ff0000">Acorn</font> 
task in module <font color="#ff0000">Program List</font>).
</li><li>Re-build in arp/warp using the ACORN phases as restraints.
</li></ul>
<p></p>

<h2>MrBUMP</h2>

<p>You have now prepared three search models based on 1v3z, and used Molrep and Phaser
to do the molecular replacement. These steps, and the initial discovery of 1v3z and
other related proteins, are automated in the program MrBUMP.</p>

<p>Depending on what you want to do, MrBUMP can make use of web-based services. The
following tutorial deliberately does <em>not</em> make use of the web, so that it
can be run anywhere. At the end of the tutorial, there are suggestions for web-based 
options. The use of a few local PDB template files also means that the tutorial
is fairly quick. Beware that a full run of MrBUMP might take longer than is 
reasonable for a tutorial.</p>

<p></p><ol>
<li>Select the <font color="#ff0000">Molecular Replacement</font> module and open the 
<font color="#ff0000">Run MrBUMP</font> task window.
</li><li>Enter a title.
</li><li>Leave <font color="#009900">Program Mode</font> 
<font color="#ff0000">Model search and Molecular Replacement</font> unchanged.
</li><li>Enter <font color="#009900">SEQ in</font> 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/hypF_Ndom.seq</font>
</li><li>Enter <font color="#009900">MTZ in</font> 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/hypF-1gxu-1gxt-HG_scaleit1.mtz</font> 
and select the labels <font color="#009900">F</font> <font color="#ff0000">FP1gxu</font>, 
<font color="#009900">SIGF</font> <font color="#ff0000">SIGFP1gxu</font> and
<font color="#009900">Free</font> <font color="#ff0000">FREE</font>
</li><li>Leave the rest of the files folder unchanged, and move to the
<font color="#009900">Template Search Options</font> folder.
</li><li>Un-check <font color="#009900">Do a FASTA search for possible template models</font>.
Instead we are going to use some known local templates.
</li><li>Un-check <font color="#009900">Update local copies of search databases</font>
</li><li>Select <font color="#009900">Multiple alignment program</font>
<font color="#ff0000">Mafft</font> if available
</li><li>Un-check all <font color="#009900">Additional search methods</font>, i.e. SCOP,
PQS and SSM
</li><li>The folder <font color="#009900">User specified search models</font> will have
opened. Because we have switched off all search options, we are required to use 
local files. Click on <font color="#ff0000">Add PDB file</font> 3 times to add 3 local
PDB files. The first file is <font color="#ff0000">$MR_TUTORIAL/data/hypF/1w2i_A.pdb</font>
and specify <font color="#009900">Chain identifier</font><font color="#ff0000">A</font>.
The second file is <font color="#ff0000">$MR_TUTORIAL/data/hypF/1v3z_B.pdb</font>
and specify <font color="#009900">Chain identifier</font><font color="#ff0000">B</font>.
The third file is <font color="#ff0000">$MR_TUTORIAL/data/hypF/2acy.pdb</font>
and specify <font color="#009900">Chain identifier</font><font color="#ff0000">A</font>.
</li><li>In the folder <font color="#009900">Search Model Preparation Options</font>,
keep the default which is to use <font color="#009900">Molrep</font>,  
<font color="#009900">Chainsaw</font> and <font color="#009900">Sculptor</font>. This means there will be 9 search models in
total. Turn one or two off to make the job quicker.
</li><li>In the folder <font color="#009900">Molecular Replacement and Refinement Options</font>,
keep <font color="#009900">Molrep</font> and switch off <font color="#009900">Phaser</font>. If you
want, you can use <font color="#009900">Phaser</font> instead of <font color="#009900">Molrep</font> or both.
</li><li>In the folder <font color="#009900">Model Building and Phase Improvement</font>, select the model building programs to try after MR and refinement. By default Buccaneer is set but depending on your installation you may be able to try ARP/wARP and c-alpha tracing with SHELXE as well. Model building can help determine if MR has been successful. 
</li><li><font color="#ff0000">Click Run -&gt; Run Now</font>
</li></ol>
<p></p>

<p>After a few minutes, have a look at the MrBUMP log file (do not wait for the
job to finish).
</p><ul>
<li>At the top, it echoes the options selected.
</li><li>Under <font color="#009900">Target Information</font>, it estimates that there
is 1 molecule in the target asymmetric unit.
</li><li>Under <font color="#009900">Template Model Search Results</font>, it lists
the three local files entered. They are named "loc0", "loc1", "loc2" for internal
use.
</li><li>Under <font color="#009900">Search Model Preparation Results</font>, details
of the Molrep, Chainsaw and Sculptor methods are given.
</li><li>Finally, the section <font color="#009900">Molecular Replacement and Refinement</font>
gives details for every MR job tried.
</li></ul>
<p></p>

<p>By default, it will finish when it finds a solution. For example, it may finish
with model loc1_B_MOLREP, which corresponds to template 1v3z_B.pdb with a search
model created with the Molrep editing features. The Rfree drops from 0.549 to
0.436 (precise numbers may vary!) indicating that the MR solution is refinable,
and likely to be correct. If you want to try all search models in MR (a good idea
unless you are in a rush), select <font color="#009900">Finish when</font>
<font color="#ff0000">all of the search models have been tried in MR</font> in
the folder <font color="#009900">Molecular Replacement and Refinement Options</font>.</p>

<p>If there are no problems accessing web-based services, then you can search for
templates rather than use local PDB files. Run as above, with the following differences:
</p><ol>
<li>In the folder <font color="#009900">Template Search Options</font>,
check <font color="#009900">Do a FASTA search for possible template models</font>.
</li><li>Check <font color="#009900">Run the FASTA search locally</font>. This refers
just to the search step - the PDB files are still downloaded from the web.
</li><li>Check all of the <font color="#009900">Additional search methods</font>, i.e. SCOP,
PQS and SSM
</li><li>Do not enter anything into the folder 
<font color="#009900">User specified search models</font>.
</li></ol>
<p></p>

<p>For comparison, here are some example results from MrBUMP (you may not get exactly the
same):</p>
<p>
<table>
<tbody><tr><th><b>PDB chain&nbsp;&nbsp;&nbsp;</b> </th>
<th><b>sequence identity&nbsp;&nbsp;&nbsp;</b> </th>
<th><b>source / release date</b>  
</th><th><b>Rfree from MrBUMP</b></th></tr>
<tr><td>1w2i_B 	 </td><td>0.310 	 </td><td>OCA - released Apr 2005   
</td><td>chainsaw 0.447     molrep 0.442</td></tr>
<tr><td>1w2i_A   </td><td>0.310 	 </td><td>OCA   
</td><td>chainsaw 0.471     molrep 0.527</td></tr>
<tr><td>1v3z_B   </td><td>0.310 	 </td><td>OCA - released Mar 2005   
</td><td>chainsaw 0.430     molrep 0.453</td></tr>
<tr><td>1v3z_A   </td><td>0.310 	 </td><td>OCA   
</td><td>chainsaw 0.474     molrep 0.470</td></tr>
<tr><td>2bje_G   </td><td>0.287 	 </td><td>OCA - released Nov 2005   
</td><td>chainsaw 0.458     molrep 0.442 </td></tr>
<tr><td>2bje_E   </td><td>0.287 	 </td><td>OCA   
</td><td>chainsaw 0.468     molrep 0.486</td></tr>
<tr><td>2bje_C   </td><td>0.287 	 </td><td>OCA   
</td><td>chainsaw 0.491     molrep 0.481</td></tr>
<tr><td>2bje_A   </td><td>0.287 	 </td><td>OCA   
</td><td>chainsaw 0.448     molrep 0.443</td></tr>
<tr><td>2bjd_B   </td><td>0.287 	 </td><td>OCA - released Nov 2005   
</td><td>chainsaw 0.468     molrep 0.529</td></tr>
<tr><td>2bjd_A   </td><td>0.287 	 </td><td>OCA   
</td><td>chainsaw 0.544     molrep 0.466</td></tr>
<tr><td>1y9o_A   </td><td>0.275 	 </td><td>OCA - released Jan 2006 (NMR)   
</td><td>(not tried)</td></tr>
<tr><td>1ulr_A   </td><td>0.286 	 </td><td>OCA - released Nov 2004   
</td><td>chainsaw 0.476     molrep 0.471</td></tr>
<tr><td>2acy_A   </td><td>0.264 	 </td><td>SSM - released Nov 1997 (authors tried)&nbsp;  
</td><td>chainsaw 0.539     molrep 0.564</td></tr>
</tbody></table>
</p>

<center><h2>Advanced tutorial (OPTIONAL)</h2></center>

<h2>Other search models for hypF</h2>

<p>Another possible search model is chain A of 1w2i. This is a different structure
of the same protein as 1v3z. Try repeating the above steps using 
<font color="#ff0000">$MR_TUTORIAL/data/hypF/1w2i_A.pdb</font> as the search model.
</p>
<p>You should find that this is more difficult! Modifying the search model using
the target sequence is now necessary. Adjusting the resolution limits also helps.</p>
<p>Check your solutions against those produced from 1v3z_B.</p>

<h2>Other structure soltuions</h2>

<p>See <a href="mr_tutorial_advanced.html">separate document</a> for 3 more example MR problems.</p>

</body></html>
