<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
<HEAD>
   <TITLE>User Guide for ClustalW Interface</TITLE>
   <META NAME="GENERATOR" CONTENT="Mozilla/3.0Gold (X11; I; IRIX 6.2 IP22) [Netscape]">
</HEAD>
<BODY>

<!-- ::INDEX_INFO::CLUSTALW INTERFACE::Supported::Coordinate file manipulations, Protein sequence file manipulations::Graphical Interface to the ClustalW Program:::::::: -->

<a name="top"></a><H1>CLUSTALW INTERFACE</H1>
<H2>Contents</H2>
<dl>
<dt><a href="#description" target="_self">Description</a></dt>
<dt><a href="#Running ClustalW" target="_self">Running ClustalW</a></dt>
<dt><a href="#Specific Options" target="_self">Specific Options</a></dt>
<dd><a href="#DNA Pairwise Alignment" target="_self">DNA Pairwise Alignment Options</a></dd>
<dd><a href="#DNA Multiple Alignment" target="_self">DNA Multiple Alignment Options</a></dd>
<dd><a href="#Protein Pairwise Alignment" target="_self">Protein Pairwise Alignment Options</a></dd>
<dd><a href="#Protein Multiple Alignment" target="_self">Protein Multiple Alignment Options</a></dd>
<dt><a href="#Other Options" target="_self">Other Options</a></dt>
<dd><a href="#Structure Alignments" target="_self">Structure Alignments Options</a></dd>
<dd><a href="#Phylogenetic Trees" target="_self">Phylogenetic Trees Options</a></dd>
<dt><a href="#Observing Output" target="_self">Observing Output</a></dd>
</dl>

<H2><a name="description"></A>Description</H2>
<p>This is an interface to allow users to run the ClustalW Multiple Sequence Alignments program. Note: The ClustalW program is not included in the CCP4 distribution. ClustalW must be installed on the system running CCP4i in order to work. 

<H2><a name="Running ClustalW">Running ClustalW</a></H2>
The interface is able to run pairwise and multiple sequence alignments for both DNA and protein sequences. The minimum information required to run the program is a file containing DNA or protein sequences that need to be aligned. This file should ideally be a plain text file in the standard FASTA format. The default settings within the program should be sufficient for most cases of DNA and Protein alignment, but there are other more complicated options available, There are options to set the output type of the results as well as options to set limits for structure alignments, phylogenetic trees and options specific for DNA and protein pairwise/multiple alignments. For full explanations of these options, please refer to the ClustalW documentation.

<H2><a name="Specific Options">Specific Options</a></H2>
When the mode of the alignment is selected - DNA Pairwise, DNA Multiple, Protein Pairwise or Protein Multiple Alignments - a folder will appear with options specific for that mode.
<H3><a name="DNA Pairwise Alignment">DNA Pairwise Alignment</a></H3>
Includes options to set the DNA weight matrix, K-TUPLE value and gap penalties, including opening and extension penalties.
<H3><a name="DNA Multiple Alignment">DNA Multiple Alignment</a></H3>
Includes options to set the DNA weight matrix, gap opening and extension penalties, transitions weighting and the gap separation penalty range. Can also specify no hydrophilic gaps and no residue-specific gaps.
<H3><a name="Protein Pairwise Alignment">Protein Pairwise Alignment</a></H3>
As DNA Pairwise Alignment options, except uses Protein Weight Matrices.
<H3><a name="Protein Multiple Alignment">Protein Multiple Alignment</a></H3>
As DNA Multiple Alignment options, except uses Protein Weight Matrices.
<br><image src="images/clustal1.jpg" height=400></image>


<H2><a name="Other Options">Other Options</a></H2>
As well as the options that are specific for the mode of alignment, there are other general options that can be selected regardless of the mode.
<H3><a name="Structure Alignments">Structure Alignments Options</a></H2>
Allows you to set gap penalties for helix core residues, strand core residues, loop regions and structure termini. Also allows you to set the number of residues inside and outside the helix or strand to be treated as termini.
<br><image src="images/clustal2.jpg" height=400></image>
<H3><a name="Phylogenetic Trees">Phylogenetic Trees Options</a></H2>
Allows you to specify the production and format of phylogenetic trees. Includes the use of Kimura's Correction and ignoring positions with gaps.
<br><image src="images/clustal3.jpg" height=400></image>
<H2><a name="Observing Output">Observing Output</a></H2>
The output of the program is written to a specified output file, but for a quick glance it is also written to a viewing window in the "Output" folder at the bottom of the interface.
<br><image src="images/clustal4.jpg" height=400></image>














