<!doctype html system "html.dtd">
<html>
<head><title>CCP4 Program Documentation</title></head>
<body  BGCOLOR="#FFFFFF">
<FONT COLOR="#FF0000"><H1>CCP4 v6.4.0 Program Documentation</H1></FONT>
<P><B><U><FONT SIZE="3">Reference:</FONT></U></B><BR>
Collaborative Computational Project, Number 4. 1994.<BR>
 &quot;The CCP4 Suite: Programs for Protein Crystallography&quot;. Acta Cryst. D50,
760-763 </P>
<p>
There is a partial list of <a href="REFERENCES.html">program references</a>
extracted from the individual program documentation.
<HR>

<P><B><U><FONT SIZE="3">Changes:</FONT></U></B><BR>
List of the <a href="CHANGESinV6_4.html">major program changes</a>
since the previous release.
<HR>

This list can also be seen in <a href="INDEX.html">alphabetical order.</a>

<p>
<FONT COLOR="#FF0000"><H2>COORDINATE FILE MANIPULATIONS</H2></FONT>
<ul>
<li><a href="bones2pdb.html">bones2pdb</a> - Make a PDB pseudo-coordinate file from a bones file  </li> 
<li><a href="chainsaw.html">chainsaw</a> - Mutate a pdb file according to an input sequence alignment  </li> 
<li><a href="clustalw.html">clustalw interface</a> - Graphical Interface to the ClustalW Program  </li> 
<li><a href="coord_format.html">coord_format</a> - fix PDB format and convert to/from other formats  </li> 
<li><a href="coordconv.html">coordconv</a> - Interconvert various coordinate formats  </li> 
<li><a href="csymmatch.html">csymmatch</a> - Use symmetry to match chains  </li> 
<li><a href="gensym.html">gensym</a> - generate sites by symmetry  </li> 
<li><a href="geomcalc.html">geomcalc</a> - molecular geometry calculations  </li> 
<li><a href="gesamt.html">gesamt</a> - structural alignment  </li> 
<li><a href="hgen.html">hgen</a> - generate hydrogen atom positions for proteins  </li> 
<li><a href="lsqkab.html">lsqkab</a> - apply various transformations to coordinate files  </li> 
<li><a href="makedict.html">makedict</a> - converts PDB file to TNT or PROTIN dictionaries and PROTIN to PDB  </li> 
<li><a href="pdb_merge.html">pdb_merge</a> - merge two coordinate files into one  </li> 
<li><a href="pdbcur.html">pdbcur</a> - various useful manipulations on coordinate files  </li> 
<li><a href="pdbset.html">pdbset</a> - various useful manipulations on coordinate files  </li> 
<li><a href="reforigin.html">reforigin</a> - apply best origin shift to PDB atom co-ords according to reference file (unsupported) </li> 
<li><a href="rwcontents.html">rwcontents</a> - Count atoms by type  </li> 
<li><a href="sortwater.html">sortwater</a> - sort waters by the protein chain to which they "belong"  </li> 
<li><a href="superpose.html">superpose</a> - structural alignment based on secondary structure matching  </li> 
<li><a href="watertidy.html">watertidy</a> - rationalise waters at the end of refinement  </li> 
<li><a href="watpeak.html">watpeak</a> - selects peaks from peakmax and puts them close to the respective protein atoms  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>DATA HARVESTING AND DATASETS</H2></FONT>
<ul>
<li><a href="cif2xml.html">cif2xml</a> - Conversion of mmCIF files to XML   </li> 
<li><a href="cross_validate.html">cross_validate</a> - Validation of harvest files for deposition   </li> 
<li><a href="dhm_tool.html">data harvesting manager</a> - Tool for managing Data Harvesting Files  </li> 
<li><a href="harvesting.html">harvesting</a> - harvesting data automatically and using datasets  </li> 
<li><a href="harvlib.html">harvlib</a> - Subroutine library for writing CIF harvest files  </li> 
<li><a href="pdb_extract.html">pdb_extract</a> - RCSB/PDB Programs for extracting harvest information from program log files  </li> 
<li><a href="r500.html">r500</a> - Checks PDB files for any problems before submission  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>DATA PROCESSING AND REDUCTION</H2></FONT>
<ul>
<li><a href="aimless.html">aimless</a> - scale together multiple observations of reflections  </li> 
<li><a href="combat.html">combat</a> - produces an MTZ file in multirecord form suitable for input to SCALA.  </li> 
<li><a href="crossec.html">crossec</a> - interpolate X-ray cross sections and compute anomalous scattering factors  </li> 
<li><a href="ctruncate.html">ctruncate</a> - Intensity to amplitude conversion  </li> 
<li><a href="detwin.html">detwin</a> - detwins merohedrally twinned data  </li> 
<li><a href="dtrek2mtz.html">dtrek2mtz</a> - converts d*trek scalemerge output into MTZ format  </li> 
<li><a href="dtrek2scala.html">dtrek2scala</a> - initial processing of intensity files from D*TREK  </li> 
<li><a href="pointless.html">pointless</a> - determine Laue group  </li> 
<li><a href="rebatch.html">rebatch</a> - alter batch numbers in an unmerged MTZ file  </li> 
<li><a href="reindex.html">reindex</a> - produces an MTZ file with h k l reindexed and/or the symmetry changed  </li> 
<li><a href="reindexing.html">reindexing</a> - information about changing indexing regime (see reindex) </li> 
<li><a href="rotaprep.html">rotaprep</a> - produces an MTZ file in multirecord form suitable for input to SCALA.  </li> 
<li><a href="scala.html">scala</a> - scale together multiple observations of reflections  </li> 
<li><a href="scalepack2mtz.html">scalepack2mtz</a> - converts merged scalepack output into MTZ format  </li> 
<li><a href="truncate.html">truncate</a> - obtain structure factor amplitudes using Truncate procedure  </li> 
<li><a href="twinning.html">twinning</a> - dealing with data from twinned crystals  </li> 
<li><a href="unique.html">unique</a> - Generate a unique list of reflections (including uniqueify) </li> 
<li><a href="wilson.html">wilson</a> - Wilson plot, absolute scale and temperature factor  </li> 
<li><a href="xia2.html">xia2</a> - Automated data reduction  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>DENSITY FITTING AND MODEL BUILDING</H2></FONT>
<ul>
<li><a href="arp_waters.html">arp_waters</a> - Automated Refinement Procedure for refining protein structures  </li> 
<li><a href="cbuccaneer.html">buccaneer</a> - Statistical Model Building  </li> 
<li><a href="cprodrg.html">prodrg</a> - generation of small molecule coordinates  </li> 
<li><a href="fffear.html">fffear</a> - map interpretation package  </li> 
<li><a href="fffear_fraglib.html">fffear fragment library</a> - map interpretation package  </li> 
<li><a href="ffjoin.html">ffjoin</a> - joining model fragments from FFFEAR  </li> 
<li><a href="rapper.html">rapper</a> - conformer modelling and building  </li> 
<li><a href="sequins.html">sequins</a> - Statistical Model Building  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>DENSITY MODIFICATION - MOLECULAR AVERAGING</H2></FONT>
<ul>
<li><a href="chltofom.html">hltofom</a> - Convert to/from Hendrickson-Lattman coefficients  </li> 
<li><a href="cmakereference.html">cmakereference</a> - Generate reference structure for pirate/buccaneer  </li> 
<li><a href="cpirate.html">pirate</a> - Statistical Phase Improvement  </li> 
<li><a href="dm.html">dm</a> - density modification package  </li> 
<li><a href="dm_ncs_averaging.html">dm_ncs_averaging</a> - dm for ncs averaging (see DM program) </li> 
<li><a href="dm_skeletonisation.html">dm_skeletonisation</a> - iterative skeletonisation using dm (see DM program) </li> 
<li><a href="dmmulti.html">dmmulti</a> - multi-xtal density modification package  </li> 
<li><a href="matthews_coef.html">matthews_coef</a> - Misha Isupov's Jiffy to calculate Matthews coefficient  </li> 
<li><a href="ncsmask.html">ncsmask</a> - averaging mask manipulation program  </li> 
<li><a href="parrot.html">parrot</a> - density modification package  </li> 
<li><a href="professs.html">professs</a> - determination of NCS operators from heavy atoms  </li> 
<li><a href="solomon.html">solomon</a> - density modification (phase improvement) by solvent flipping  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>EXPERIMENTAL PHASING</H2></FONT>
<ul>
<li><a href="crank.html">crank</a> - Automated structure solution for experimental phasing  </li> 
<li><a href="rantan.html">rantan</a> - Direct Method module for the determination of heavy atom positions  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>MAP CALCULATION AND MANIPULATION</H2></FONT>
<ul>
<li><a href="maprot.html">maprot</a> - map skewing, interpolating, rotating, averaging and correlation masking program  </li> 
<li><a href="extends.html">extends</a> - Extend Fourier maps and compute standard uncertainty of electron density  </li> 
<li><a href="fft.html">fft</a> - fast Fourier transform  </li> 
<li><a href="mama2ccp4.html">mama2ccp4</a> - Convert between `mama' and Cambridge/CCP4 map formats  </li> 
<li><a href="map2fs.html">map2fs</a> - convert CCP4 map to XtalView fsfour format  </li> 
<li><a href="mapdump.html">mapdump</a> - print a dump of sections of a map file  </li> 
<li><a href="mapmask.html">mapmask</a> - map/mask extend program  </li> 
<li><a href="maprot.html">maprot</a> - map skewing, interpolating, rotating, averaging and correlation masking program  </li> 
<li><a href="mapsig.html">mapsig</a> - print statistics on signal/noise for translation function map  </li> 
<li><a href="maptona4.html">maptona4</a> - Convert binary map file to and from na4 ascii format  </li> 
<li><a href="omit.html">omit</a> - program to calculate omit-maps according to Bhat procedure  </li> 
<li><a href="overlapmap.html">overlapmap</a> - calculates the average of two maps  </li> 
<li><a href="peakmax.html">peakmax</a> - search for peaks in the electron density map  </li> 
<li><a href="sigmaa.html">sigmaa</a> - Improved Fourier coefficients using calculated phases  </li> 
<li><a href="xdlmapman.html">xdlmapman</a> - manipulation, analysis and reformatting of electron density maps (X-windows tool) </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>MAXIMUM LIKELIHOOD PHASING</H2></FONT>
<ul>
<li><a href="phaser.html">phaser-2.5.0</a> - Molecular Replacement, SAD Phasing, Anisotropy Correction, Cell Content Analysis, Translational NCSand Twin Analysis  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>MODEL ANALYSIS</H2></FONT>
<ul>
<li><a href="act.html">act</a> - analyse coordinates  </li> 
<li><a href="areaimol.html">areaimol</a> - Analyse solvent accessible areas (supported) </li> 
<li><a href="baverage.html">baverage</a> - averages B over main and side chain atoms  </li> 
<li><a href="cavenv.html">cavenv</a> - Visualise cavities in macromolecular structures  </li> 
<li><a href="contact.html">contact</a> - computes various types of contacts in protein structures  </li> 
<li><a href="distang.html">distang</a> - Distances and angles calculation   </li> 
<li><a href="dyndom.html">dyndom</a> - determine dynamic domains when two conformations are available  </li> 
<li><a href="ncont.html">ncont</a> - computes various types of contacts in protein structures  </li> 
<li><a href="polypose.html">polypose</a> - program for superimposing many multi-domain structures  </li> 
<li><a href="procheck.html">procheck</a> - programs to check the Stereochemical Quality of Protein Structures  </li> 
<li><a href="rampage.html">rampage</a> - Ramachandran plots  </li> 
<li><a href="rotamer.html">rotamer</a> - List amino acids whose side chain torsion angles deviate from the Penultimate Rotamer Library  </li> 
<li><a href="rotamer_table.html">rotamer table</a> - rotamer tables used in the `rotamer' program  </li> 
<li><a href="sc.html">sc</a> - analyse shape complementarity (supported) </li> 
<li><a href="seqwt.html">seqwt</a> - Calculate molecular weight from sequence  </li> 
<li><a href="sfcheck.html">sfcheck</a> - program for assessing the agreement between the atomic model and X-ray data  </li> 
<li><a href="surface.html">surface</a> - surface accessibility program and for preparing input file to program volume  </li> 
<li><a href="topp.html">topp</a> - a topological comparison program  </li> 
<li><a href="volume.html">volume</a> - polyhedral volume around selected atoms  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>MODEL REFINEMENT</H2></FONT>
<ul>
<li><a href="anisoanl.html">anisoanl</a> - analyses of anisotropic displacement parameters   </li> 
<li><a href="arp_waters.html">arp_waters</a> - Automated Refinement Procedure for refining protein structures  </li> 
<li><a href="dimple.html">dimple</a> - automated difference map calculation  </li> 
<li><a href="intro_mon_lib.html">intro_mon_lib</a> - description of monomer library used in CCP4  </li> 
<li><a href="lib_list.html">lib_list</a> - contents of multi-purpose monomer dictionary used by REFMAC  </li> 
<li><a href="libcheck.html">libcheck</a> - monomer library management program  </li> 
<li><a href="mon_lib.html">mon_lib</a> - description of multi-purpose monomer dictionary used by REFMAC  </li> 
<li><a href="rdent.html">rdent</a> - Create dictionary entries for Restrain from PDB file  </li> 
<li><a href="refmac5.html">refmac5</a> - macromolecular refinement program  </li> 
<li><a href="restrain.html">restrain</a> - macromolecular refinement program  </li> 
<li><a href="sfall.html">sfall</a> - Structure factor calculation and X-ray refinement using forward and reverse FFT  </li> 
<li><a href="sketcher.html">sketcher</a> - monomer library sketcher  </li> 
<li><a href="tlsanl.html">tlsanl</a> - analysis of TLS tensors and derived anisotropic U factors  </li> 
<li><a href="tlsextract.html">tlsextract</a> - extract TLS group description from a PDB file  </li> 
<li><a href="watncs.html">watncs</a> - Pick waters which follow NCS and sort out to NCS asymmetric unit  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>MOLECULAR REPLACEMENT</H2></FONT>
<ul>
<li><a href="almn.html">almn</a> - calculates rotation function overlap values using FFT techniques (alternative to AMORE) </li> 
<li><a href="amore.html">amore</a> - Jorge Navaza's state-of-the-art molecular replacement package  </li> 
<li><a href="beast.html">beast</a> - Likelihood-based molecular replacement  </li> 
<li><a href="bulk.html">bulk</a> - bulk solvent correction for translation search and rigid body refinement steps of AMoRe   </li> 
<li><a href="ecalc.html">ecalc</a> - calculate normalised structure amplitudes  </li> 
<li><a href="fsearch.html">fsearch</a> - 6-d molecular replacement (envelope) search  </li> 
<li><a href="getax.html">getax</a> - real space correlation search  </li> 
<li><a href="molrep.html">molrep</a> - automated program for molecular replacement  </li> 
<li><a href="mrbump_doc.html">mrbump</a> - automated search model generation and automated molecular replacement  </li> 
<li><a href="mrbump_doc.html">mrbump</a> - automated search model generation and automated molecular replacement  </li> 
<li><a href="polarrfn.html">polarrfn</a> - fast rotation function that works in polar angles  </li> 
<li><a href="rfcorr.html">rfcorr</a> - Analysis of correlations between cross- and self-Rotation functions  </li> 
<li><a href="rotmat.html">rotmat</a> - interconverts CCP4/MERLOT/X-PLOR rotation angles  </li> 
<li><a href="rsearch.html">rsearch</a> - R-factor and correlation coefficient between Fcalc and Fobs  </li> 
<li><a href="seqwt.html">seqwt</a> - Calculate molecular weight from sequence  </li> 
<li><a href="tffc.html">tffc</a> - Translation Function Fourier Coefficients  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>MTZ MANIPULATIONS / CONVERSIONS</H2></FONT>
<ul>
<li><a href="cad.html">cad</a> - Collect and sort crystallographic reflection data from several files  </li> 
<li><a href="cif2mtz.html">cif2mtz</a> - Convert an mmCIF reflection file to MTZ format  </li> 
<li><a href="convert2mtz.html">convert2mtz</a> - CNS to MTZ conversion  </li> 
<li><a href="cphasematch.html">cphasematch</a> - Calculate agreement between phase sets  </li> 
<li><a href="ecalc.html">ecalc</a> - calculate normalised structure amplitudes  </li> 
<li><a href="f2mtz.html">f2mtz</a> - Convert a formatted reflection file to MTZ format  </li> 
<li><a href="freerflag.html">freerflag</a> - tags each reflection in an MTZ file with a flag for cross-validation  </li> 
<li><a href="freerunique.html">freerunique</a> - convert FreeRflags between CCP4 and other formats (see freerflag,uniqueify,mtz2various,f2mtz) </li> 
<li><a href="mtz2cif.html">mtz2cif</a> - produce mmCIF structure factor file suitable for deposition  </li> 
<li><a href="mtz2various.html">mtz2various</a> - produces reflexion file for MULTAN, SHELX, TNT, X-PLOR/CNS, pseudo-SCALEPACK, XtalView, mmCIF or other ascii format  </li> 
<li><a href="mtzMADmod.html">mtzMADmod</a> - Generate F+/F- or F/D from other for anomalous data  </li> 
<li><a href="mtzdump.html">mtzdump</a> - dump data from an MTZ reflection data file  </li> 
<li><a href="mtzmnf.html">mtzmnf</a> - Identify missing data entries in an MTZ file and replace with a Missing Number Flag (MNF)  </li> 
<li><a href="mtztona4.html">mtztona4</a> - interconvert MTZ reflection file and ASCII format  </li> 
<li><a href="mtzutils.html">mtzutils</a> - Reflection data files utility program  </li> 
<li><a href="sftools.html">sftools</a> - reflection data file utility program  </li> 
<li><a href="sortmtz.html">sortmtz</a> - Sort a MTZ reflection data file  </li> 
<li><a href="xdldataman.html">xdldataman</a> - manipulation, analysis and reformatting of reflection files (X-windows tool) </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>MULTIMER ANALYSIS</H2></FONT>
<ul>
<li><a href="pisa.html">pisa</a> - Protein interfaces, surfaces and assemblies  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>MULTIPLE ISOMORPHOUS REPLACEMENT</H2></FONT>
<ul>
<li><a href="abs.html">abs</a> - determine the absolute configuration (hand) of the heavy atom substructure  </li> 
<li><a href="afro.html">afro</a> - multivariate substructure factor amplitude estimation for SAD/MAD and SIRAS  </li> 
<li><a href="bp3.html">bp3</a> - multivariate likelihood substructure refinement and phasing of S/MIR(AS) and/or S/MAD  </li> 
<li><a href="findncs.html">findncs</a> - detect NCS operations automatically from heavy atom sites  </li> 
<li><a href="gcx.html">gcx</a> - support program to generate crank XML and run crank via a script  </li> 
<li><a href="mlphare.html">mlphare</a> - maximum likelihood heavy atom refinement and phase calculation  </li> 
<li><a href="oasis.html">oasis</a> - A program for breaking phase ambiguity in OAS or SIR  </li> 
<li><a href="peakmax.html">peakmax</a> - search for peaks in the electron density map  </li> 
<li><a href="rsps.html">rsps</a> - heavy atom positions from derivative difference Patterson maps  </li> 
<li><a href="sapi.html">sapi</a> - heavy atom site location  </li> 
<li><a href="vecref.html">vecref</a> - Vector-space refinement of heavy atom sites in isomorphous derivatives  </li> 
<li><a href="vecsum.html">vecsum</a> - program to deconvolute a Patterson function and solve the structure (unsupported) </li> 
<li><a href="vectors.html">vectors</a> - generates Patterson vectors from atomic coordinates  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>MULTIWAVELENGTH ANOMALOUS DISPERSION</H2></FONT>
<ul>
<li><a href="abs.html">abs</a> - determine the absolute configuration (hand) of the heavy atom substructure  </li> 
<li><a href="afro.html">afro</a> - multivariate substructure factor amplitude estimation for SAD/MAD and SIRAS  </li> 
<li><a href="bp3.html">bp3</a> - multivariate likelihood substructure refinement and phasing of S/MIR(AS) and/or S/MAD  </li> 
<li><a href="crossec.html">crossec</a> - interpolate X-ray cross sections and compute anomalous scattering factors  </li> 
<li><a href="gcx.html">gcx</a> - support program to generate crank XML and run crank via a script  </li> 
<li><a href="mlphare.html">mlphare</a> - maximum likelihood heavy atom refinement and phase calculation  </li> 
<li><a href="oasis.html">oasis</a> - A program for breaking phase ambiguity in OAS or SIR  </li> 
<li><a href="revise.html">revise</a> - estimates optimised value of the normalised anomalous scattering using MAD data  </li> 
<li><a href="sapi.html">sapi</a> - heavy atom site location  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>NATIVE PROTEIN DATA AT ATOMIC RESOLUTION</H2></FONT>
<ul>
<li><a href="acorn.html">acorn</a> -  Molecular Replacement, Sayre Equation and Dynamic Density Modification for the determination of a protein structure  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>OTHERS/JIFFIES</H2></FONT>
<ul>
<li><a href="phistats.html">phistats</a> - Analysis of agreement between phase sets, and checking it against weighting factors  </li> 
<li><a href="stereo.html">stereo</a> - Extract coordinates from stereo diagrams  </li> 
<li><a href="stgrid.html">stgrid</a> - Generate plot to measure angular coordinates on a stereographic projection from polarrfn  </li> 
<li><a href="stnet.html">stnet</a> - Generate plot to measure angles between points on a stereographic projection from polarrfn  </li> 
<li><a href="tracer.html">tracer</a> - Lattice TRAnsformation and CEll Reduction  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>PLOTTING / GRAPHICS</H2></FONT>
<ul>
<li><a href="AstexViewer.html">openastexviewer</a> - Java program for display molecular structures and electron density maps  </li> 
<li><a href="ccp4mapwish.html">ccp4mapwish</a> - custom wish interpreter required for MapSlicer  </li> 
<li><a href="hklplot.html">hklplot</a> - plots a precession photo from an HKL data file  </li> 
<li><a href="hklview.html">hklview</a> - displays zones of reciprocal space as pseudo-precession images (X-windows program) </li> 
<li><a href="ipdisp.html">ipdisp</a> - displays images from a variety of (crystallographic) sources (X-Windows tool) </li> 
<li><a href="loggraph.html">loggraph</a> - viewer for CCP4 formatted `log' files  </li> 
<li><a href="mapslicer.html">mapslicer</a> - interactive section viewer for CCP4 map files  </li> 
<li><a href="npo.html">npo</a> - Molecule and map plotting  </li> 
<li><a href="pltdev.html">pltdev</a> - convert Plot84 meta-files to PostScript, Tektronix or HPGL  </li> 
<li><a href="rasmol.html">rasmol</a> - Molecular Visualisation Program (X-windows tool) </li> 
<li><a href="topdraw.html">topdraw</a> - Sketchpad for protein topology diagrams (supported) </li> 
<li><a href="xloggraph.html">xloggraph</a> - a viewer of specially formatted `log' files (X-windows tool) </li> 
<li><a href="xplot84driver.html">xplot84driver</a> - a viewer for Plot84 meta files (X-windows tool) </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>PROTEIN SEQUENCE FILE MANIPULATIONS</H2></FONT>
<ul>
<li><a href="clustalw.html">clustalw interface</a> - Graphical Interface to the ClustalW Program  </li> 
<li><a href="get_prot.html">import/edit protein sequence</a> - Import and edit protein sequences  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>ROTATION MATRICES</H2></FONT>
<ul>
<li><a href="rotationmatrices.html">rotationmatrices</a> - on Eulerian angles, polar angles and direction cosines, and orthogonalisation codes  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>SCALING DATA</H2></FONT>
<ul>
<li><a href="fhscal.html">fhscal</a> - Scaling of isomorphous derivative data using Kraut's method (see also SCALEIT) </li> 
<li><a href="icoefl.html">icoefl</a> - vectorially combined scaling of Fobs (Iobs) with partial Fc's  </li> 
<li><a href="rstats.html">rstats</a> - scale together two sets of F's  </li> 
<li><a href="scalechoose.html">scalechoose</a> - information about choice of scaling program (see reindex) </li> 
<li><a href="scaleit.html">scaleit</a> - derivative to native scaling (see also FHSCAL) </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>SYMMETRY UTILITIES</H2></FONT>
<ul>
<li><a href="symconv.html">symconv</a> - Fetch and convert symmetry and spacegroup information  </li> 
</ul>
<p>
<FONT COLOR="#FF0000"><H2>VALIDATION AND DEPOSITION</H2></FONT>
<ul>
<li><a href="dhm_tool.html">data harvesting manager</a> - Tool for managing Data Harvesting Files  </li> 
<li><a href="pdb_extract.html">pdb_extract</a> - RCSB/PDB Programs for extracting harvest information from program log files  </li> 
<li><a href="r500.html">r500</a> - Checks PDB files for any problems before submission  </li> 
</ul>
