<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
<HEAD>
   <TITLE>User Guide for 'Import/Edit Protein Sequence'</TITLE>
   <META NAME="GENERATOR" CONTENT="Mozilla/3.0Gold (X11; I; IRIX 6.2 IP22) [Netscape]">
</HEAD>
<BODY>

<!-- ::INDEX_INFO::IMPORT/EDIT PROTEIN SEQUENCE::Supported::Protein sequence file manipulations::Import and edit protein sequences:::::::: -->


<a name="top"></a><H1>IMPORT/EDIT PROTEIN SEQUENCE</H1>
<br />
<H2>Contents</H2>
<dl>
<dt><a href="#description" target="_self">Description</a></dt>
<dt><a href="#Viewing Protein Sequences" target="_self">Viewing Protein
Sequences</a></dt>
<dt><a href="#Editing Facilities" target="_self">Editing Facilities</a></dt>
<dd><a href="#Deletion" target="_self">Deletion</a></dd>
<dd><a href="#Mutation" target="_self">Mutation</a></dd>
<dd><a href="#Insertion" target="_self">Insertion</a></dd>
<dd><a href="#Undo" target="_self">Undo</a></dd>
<dd><a href="#Calculate Molecular Weight" target="_self">Calculate Molecular Weight</a></dd>
<dt><a href="#Output Files" target="_self">Output Files</a></dt>
<dt><a href="#Changing the external database" target="_self">Changing the
external database</a></dt>
</dl>

<H2><a name="description"></A>Description</H2>

<P>This task reads in a protein sequence in Swissprot format, either
stored locally or when downloaded from a specified database.  This
sequence can then be edited for future use. There is also a function which calculates the molecular weight of a protein.</P>
<p>The <font color="blue">"Select Database"</font> menu allows you to choose which database you
wish to obtain a protein sequence - either locally, or from an
external database.  The default external database is the Swissprot
database located at the <a href="http://www.ebi.ac.uk">EBI (European Bioinformatics Institute),</a>
although this can be changed to any database specified <a href="#Changing the external database"
target="_self">(See "Changing the external database").</a>
<br /><br /><br />

<p><a name="Viewing Protein Sequences"></a><H2>Viewing Protein Sequences</H2>
<P>If the database selected is an external database, such as the EBI/SwissProt database, specify the accession code in the <font color="blue">"Type Protein
Code"</font> box and then press the <font color="blue">"Import Sequence"</font> button.  A message will
appear showing an attempt to connect to the external database
specified.  If the connection is successful, the sequence of this
protein will appear in the <font color="blue">"Sequence Only"</font> viewing window.
<p>However, if you wish to view the full details of the selected protein, this can be
seen when the <font color="blue">"View Full Sequence Entry"</font> viewing window is opened.
<br><image src="images/getprot1.jpg" height=500></image>


<p><a name="Editing Facilities"></a><H2>Editing Facilities</H2>
<p>The interface has the ability to manipulate the protein sequence in
the <font color="blue">"Sequence Only"</font> viewing window in three
ways: <a href="#Deletion" target="_self">Deletion</a>, <a
href="#Mutation" target="_self">Mutation</a> and <a href="#Insertion"
target="_self">Insertion.</a>  There is also a <a href="#Undo"
target="_self">Undo</a> function available. A record of all editing functions carried out appears in the textframe in the "Editing" folder.

<p><a name="Deletion"></a><H3>Deletion</H3>

<p>Before any of the editing functions can be carried out, a section
of the protein sequence must be highlighted.  To do this, enter in the
<font color="blue">"Search For..."</font> box the sequence within the protein that is to be
manipulated.  If this substring is located within the protein sequence
in view, then it will be highlighted in red when the <font color="blue">"Search For..."</font>
button is pressed.
<br><image src="images/getprot2.jpg" width=550></image>

<P>While any section of the protein sequence is highlighted, editing
can be carried out.  So, by pressing the <font color="blue">"Delete"</font> button, the
previously highlighted section of the protein sequence is deleted and
a record of this deletion is shown in the viewing window at the bottom
of the <font color="blue">"Editing"</font> window.


<p><a name="Mutation"></a><H3>Mutation</H3>
<p>The "Mutation" function replaces a highlighted section of the
protein sequence with a specified section.  In the example below, the
first five residues in the protein sequence are replaced by specifying
the mutation sequence and pressing the <font color="blue">"Mutate
into..."</font> button.

<br><img src="images/getprot3.jpg" height=550></img>

<p><a name="Insertion"></a><H3>Insertion</H3>
<p>The "Insertion" function allows the insertion of residues <font
color="blue">AFTER</font> a highlighted section.  In the example
below, "YYYYY" is to be inserted after "XXXXX", the first five
residues in the sequence.  Search for "XXXXX" so it is highlighted,
specify the insertion of "YYYYY" and press the <font
color="blue">"Insert"</font> button.

<p><a name="Undo"></a><H3>Undo</H3>
<p>The <font color="blue">"Undo"</font> function allows the user to
undo the last editing function carried out.  However, like the other
editing functions, this can only be carried out if the search string
of the last editing function is highlighted.  Otherwise, unwanted
changes may occur to the protein sequence.


<p><a name="Calculating Molecular Weight"></a><H3>Calculating Molecular Weight of Protein</H3>
<p> The <font color="blue">"Calculate Molecular Weight"</font> button located in the "Sequence Only" folder allows the user to calculate the molecular weight of the last protein in the "Sequence Only" viewing window. The value for the molecular mass of the protein appears in the textframe in the "Editing" folder, with the amino acid composition of the protein appearing in the shell where CCP4i was started.
<br><image src="images/getprot4.jpg" height=550></image>

<p><a name="Output Files"></a><H2>Output Files</H2>
<p>Various files are created while running a job in "Import/Edit
Protein Sequence".  They are as follows:
<dl>
<dt><font color="blue">*:</font> A copy of the edited sequence, with the filename specified by the
user and stored in the specified Project directory.</dt>
<dt><font color="blue">*. log:</font> Logfile, usually kept in same specified directory and saved
file.  Shows any problems that might arise during download of the
protein sequence.</dt>
<dt><font color="blue">*. ent:</font> This is a copy of the full
SwissProt file of the sequence downloaded.  This file does not change
if editing is carried out.</dt>
<dt><font color="blue">*. chg:</font> This file records a list of the
changes carried out to the protein sequence.</dt>
</dl>

<p><a name="Changing the external database"></a><H2>Changing the external
database</H2>
<p>The database from which a SwissProt file is obtained is
configurable, i.e. the database from which the protein sequence is
obtained can be changed.  This is done by using the "Configure
Interface" window in the <font color="blue">"System Administration"</font> toolbar, located in
the main CCP4I window and changing the <font color="blue">"Command to retrieve sequence from
database"</font> line.
<p>Change this URL to specify where Swissprot files may be downloaded
from. Example command lines are:

<dl>
<dt><font color="blue">http://www.ebi.ac.uk.cgi-bin/swissfetch</font> which is default</dt>
<dt><font color="blue">http://ca.expasy.org/cgi-bin/get-sprot-entry</font></dt>
</dl>



<p><a href="#top" target="_self">Return to top of page</a>
</BODY>
</HTML>

















